#### CSCE 670 :: Information Storage and Retrieval :: Texas A&M University :: Spring 2019


# Homework 3:   Collaborative Filtering + Matrix Factorization

### 100 points [6% of your final grade]

### Due: Wednesday, March 27, 2019 by 11:59pm

*Goals of this homework:* Understand how collaborative filtering and matrix factorization works. Explore different methods for real-world recommendation senarios.

*Submission instructions (eCampus):* To submit your homework, rename this notebook as `UIN_hw3.ipynb`. For example, my homework submission would be something like `555001234_hw3.ipynb`. Submit this notebook via eCampus (look for the homework 3 assignment there). Your notebook should be completely self-contained, with the results visible in the notebook. We should not have to run any code from the command line, nor should we have to run your code within the notebook (though we reserve the right to do so). So please run all the cells for us, and then submit.

*Late submission policy:* For this homework, you may use as many late days as you like (up to the 5 total allotted to you).

*Collaboration policy:* You are expected to complete each homework independently. Your solution should be written by you without the direct aid or help of anyone else. However, we believe that collaboration and team work are important for facilitating learning, so we encourage you to discuss problems and general problem approaches (but not actual solutions) with your classmates. You may post on Piazza, search StackOverflow, etc. But if you do get help in this way, you must inform us by **filling out the Collaboration Declarations at the bottom of this notebook**. 

*Example: I found helpful code on stackoverflow at https://stackoverflow.com/questions/11764539/writing-fizzbuzz that helped me solve Problem 2.*

The basic rule is that no student should explicitly share a solution with another student (and thereby circumvent the basic learning process), but it is okay to share general approaches, directions, and so on. If you feel like you have an issue that needs clarification, feel free to contact either me or the TA.

# Part 1. Collaborative Filtering (50 points)

In this part, you will implement collaborative filtering on the [MovieLens Latest Datasets](http://files.grouplens.org/datasets/movielens/ml-latest-small-README.html). After removing users who left less than 20 ratings and movies with less than 20 ratings, the provided dataset has only ~1,200 items and ~500 users. You can also check the title and genres of each movie in *movies_info.csv*.

As background, read [Collaborative Filtering Recommender Systems](http://files.grouplens.org/papers/FnT%20CF%20Recsys%20Survey.pdf) and refer to the course slides `week06rec.pdf` for collaborative filtering.

## 1.1 Load the Data

Please download the dataset from Piazza. There are about 65,000 ratings in total. We split the rating data into two set. You will train with 70% of the data (in *train_movie.csv*) and test on the remaining 30% of data (in *test_movie.csv*). Each of train and test files has lines having this format: UserID, MovieID, Rating. 

First you will need to load the data and store it with any structure you like. Please report the numbers of unique users and movies in the dataset. 

In [103]:
import csv
import numpy as np
import math
from sklearn.decomposition import NMF
from copy import copy

In [ ]:
# movies and users in each of train and test sets.
# Your Code Here...
# use a user dictionary, key => user  value => another dictionary
# dict.get(i).get(j) =. user i's rating on movie j
def load_data(fileName):
    user_dict = {}
    movie_dict = {}
    with open(fileName) as csv_file:
        csv_reader = csv.reader(csv_file, delimiter = ',')
        for counter, row in enumerate(csv_reader):
            if counter > 0: 
                user_temp = {int(row[1]): int(row[2])}
                movie_temp = {int(row[0]): int(row[2])}
                if int(row[0]) not in user_dict:
                    user_dict[int(row[0])] = {}
                if int(row[1]) not in movie_dict:
                    movie_dict[int(row[1])] = {}
                user_dict.get(int(row[0]))[int(row[1])] = int(row[2])
                movie_dict.get(int(row[1]))[int(row[0])] = int(row[2])
    return user_dict, movie_dict, counter
            

In [ ]:
def build_matrix(user_dict, movie_dict):
    # build the matrix, row represent users and col represent movies
    # need to notice that matrix[i][j] = 0 doesn't necessarily mean user i didn't rate movie j, maybe user i gives
    # movies j a 0
    res = np.zeros((len(user_dict), len(movie_dict)))
    for user, movies in user_dict.items():
        for movie in movies:
            res[user][movie] = user_dict.get(user).get(movie)
    return res



In [ ]:
train_user_dict, train_movie_dict, train_counter = load_data("train_movie.csv")
test_user_dict, test_movie_dict, test_counter = load_data("test_movie.csv")

In [100]:
train_matrix = build_matrix(train_user_dict, train_movie_dict)
test_matrix = build_matrix(test_user_dict, test_movie_dict)

In [ ]:
USER_NUM = len(train_user_dict)
MOVIE_NUM = len(train_movie_dict)

In [177]:
print(len(train_user_dict), len(train_movie_dict))
print(len(test_user_dict), len(test_movie_dict))

541 1211
541 1211


## 1.2 Implement User-based Collaborative Filtering

In this part, you will implement the basic User–User Collaborative Filtering algorithm introduced in the class. Given the ratings by each user, you are going to try different methods in calculating the similarities between users. You will use equation `(c)` in `week06rec.pdf` (Page 40) to aggregate ratings. Set k = 0.05. Just consider all users as neighbors. That is, while predicting how user $u$ will rate item $i$, $\widehat{C}$ includes all users who have ratings on i in the training set.

*For this memory-based algorithm, you can only rely on the ratings in training set to predict for the testing set. That is, you assume that you don't know any ratings information in the test set except that when you evalaute your model.*

### Cosine Similarity

First, you will try to calculate the similarity between users with cosine similary following the equation on page 16 of [Collaborative Filtering Recommender Systems](http://files.grouplens.org/papers/FnT%20CF%20Recsys%20Survey.pdf). And then you need to predict the rating for each (user, movie) tuple in the test set. *Note: you don't need to subtract user mean baseline from the ratings prior to computing the similarity.*

In [106]:
# Your Code Here...
# Predict for test set

def cos_similar(matrix, id_one, id_two):
    # given two users, calculate the cosine similarity
    a = np.dot(matrix[id_one], matrix[id_two])
    b = sum(np.power(matrix[id_one], 2))
    c = sum(np.power(matrix[id_two], 2))
    return a / (math.sqrt(b) * math.sqrt(c))

def predict(user, movie, matrix, train_user_dict, mean_dict):
    # aggregate the result using cosine similarity
    aggregate = 0
    for i in range(USER_NUM):
        if (i is not user) and contains(i, movie, train_user_dict):
            aggregate += (cos_similar(matrix, i, user) * (matrix[i][movie] - mean_dict.get(i)))
    return aggregate * 0.05 
                          
def average(user, user_dict):   
    # given a user, calculate the average rating he has on all movies
    total = 0
    movies = user_dict.get(user)
    for movie in movies:
        total += movies.get(movie)
    return total / len(movies)
    

def mean_map(user_dict):
    # build a dict that map each user with his average score
    res = {}
    total = 0
    for user, movies in user_dict.items():
        total = 0
        for movie in movies:
            total += movies.get(movie)
        res[user] = total / len(movies)
    return res;


def contains(user, movie, user_dict):
    # given a user and a movie, return true if this user rated this movie before
    if movie in user_dict.get(user):
        return True;
    return False


In [24]:
# build the mean dict
mean_dict = mean_map(train_user_dict)

In [25]:
# results for cosine
result = {}
for user, movies in test_user_dict.items():
    for movie in movies:
        temp = mean_dict.get(user) + predict(user, movie, train_matrix, train_user_dict, mean_dict)
        result[(user, movie)] = temp

### Evaluation

You should evaluate your predictions using Mean Absolute Error and Root Mean Squared Error. 

In [138]:
def evaluation(result, test_user_dict, test_counter):
    # given a result, evaluate the result
    MAE = 0
    RMSE = 0
    for user, movies in test_user_dict.items():
        for movie in movies:
            MAE += abs(movies.get(movie) - result.get((user, movie)))
            RMSE += pow((movies.get(movie) - result.get((user, movie))), 2)
    MAE /= test_counter
    RMSE = math.sqrt(RMSE / test_counter)
    return MAE, RMSE

In [178]:
# Your Code Here...
# Report Mean Absolute Error and Root Mean Squared Error for test set
cos_MAE, cos_RMSE = evaluation(result, test_user_dict, test_counter) 
print("User-based Collaborative Filtering with Cosine Similarity: ")
print("MAE: ", cos_MAE, "  RMSE: ", cos_RMSE)  # 0.6903937198616398 0.8991489107420851


User-based Collaborative Filtering with Cosine Similarity: 
MAE:  0.6903937198616398   RMSE:  0.8991489107420851


### Pearson correlation

Then, you will try to calculate the similarity between users with pearson correlation following `week06rec.pdf` (Page 37). And then you need to predict the rating for each (user, movie) tuple in the test set.

In [180]:
# Your Code Here...
# Predict for test set


def pearson_correlation(user_dict, id_one, id_two, mean_dict):
    # given two users, calculate their pearson correlation
    a = 0
    b = 0
    c = 0
    for movie in user_dict.get(id_one):
        if movie in user_dict.get(id_two):
            a += (user_dict.get(id_one).get(movie) - mean_dict.get(id_one)) * (user_dict.get(id_two).get(movie) - mean_dict.get(id_two))
            b += pow(user_dict.get(id_one).get(movie) - mean_dict.get(id_one), 2)
            c += pow(user_dict.get(id_two).get(movie) - mean_dict.get(id_two), 2)
    if b * c == 0:
        return 0
    return a / math.sqrt(b * c)


def pearson_predict(user, movie, train_user_dict, mean_dict):
    # aggregate the result
    aggregate = 0
    for i in range(len(train_user_dict)):
        if (i is not user) and contains(i, movie, train_user_dict):
            aggregate += (pearson_correlation(train_user_dict, user, i, mean_dict) * (train_user_dict.get(i).get(movie) - mean_dict.get(i)))
    return aggregate * 0.05 
    


In [181]:
pearson_result = {}
for user, movies in test_user_dict.items():
    for movie in movies:
        temp = mean_dict.get(user) + pearson_predict(user, movie, train_user_dict, mean_dict)
        pearson_result[(user, movie)] = temp

### Evaluation

You should evaluate your predictions using Mean Absolute Error and Root Mean Squared Error. 

In [182]:
# Your Code Here...
# Report Mean Absolute Error and Root Mean Squared Error for test

pearson_MAE, pearson_RMSE = evaluation(pearson_result, test_user_dict, test_counter)   
print("User-based Collaborative Filtering with Pearson Similarity: ")
print("MAE: ", pearson_MAE, "  RMSE: ", pearson_RMSE)  # 0.6917506759527966 0.8983386235087089

User-based Collaborative Filtering with Pearson Similarity: 
MAE:  0.6917506759527966   RMSE:  0.8983386235087089


### Pearson correlation (varying the threshold)

In [Collaborative Filtering Recommender Systems](http://files.grouplens.org/papers/FnT%20CF%20Recsys%20Survey.pdf), they observe that: 

> Pearson correlation suffers from computing high similarity
between users with few ratings in common. This can be alleviated by setting a threshold on the number of co-rated items
necessary for full agreement (correlation of 1) and scaling the
similarity when the number of co-rated items falls below this
threshold.

So now revise your Pearson to consider a threshold. Try several values and report for one that you think is appropriate.

In [196]:
# Your Code Here...
# Predict for test set
def pearson_threshold(user_dict, id_one, id_two, mean_dict):
    a = 0
    b = 0
    c = 0
    common = 0
    for movie in user_dict.get(id_one):
        if movie in user_dict.get(id_two):
            common += 1
            a += (user_dict.get(id_one).get(movie) - mean_dict.get(id_one)) * (user_dict.get(id_two).get(movie) - mean_dict.get(id_two))
            b += pow(user_dict.get(id_one).get(movie) - mean_dict.get(id_one), 2)
            c += pow(user_dict.get(id_two).get(movie) - mean_dict.get(id_two), 2)
    if b * c == 0:
        return 0;
    elif common > 25:
        return a / math.sqrt(b * c)
    return (common / 25) * a / math.sqrt(b * c)

def pearson_threshold_predict(user, movie, train_user_dict, mean_dict):
    aggregate = 0
    for i in range(len(train_user_dict)):
        if (i is not user) and contains(i, movie, train_user_dict):
            aggregate += (pearson_threshold(train_user_dict, user, i, mean_dict) * (train_user_dict.get(i).get(movie) - mean_dict.get(i)))
    return aggregate * 0.05 


In [197]:
# Your Code Here...
# Report Mean Absolute Error and Root Mean Squared Error for test

threshold_result = {}
for user, movies in test_user_dict.items():
    for movie in movies:
        temp = mean_dict.get(user) + pearson_threshold_predict(user, movie, train_user_dict, mean_dict)
        threshold_result[(user, movie)] = temp

### Evaluation

You should evaluate your predictions using Mean Absolute Error and Root Mean Squared Error. 

In [198]:
threshold_MAE, threshold_RMSE = evaluation(threshold_result, test_user_dict, test_counter)  
print("User-based Collaborative Filtering with Pearson Similarity(Varying the Threshold): ")
print("MAE: ", threshold_MAE, "  RMSE: ", threshold_RMSE)  # 0.6917506759527966 0.8983386235087089

User-based Collaborative Filtering with Pearson Similarity(Varying the Threshold): 
MAE:  0.6883253384684946   RMSE:  0.8936574572009663


What do you observe? How different are your results for the original Pearson Correlation approach vs. the thresholded version vs. the Cosine Similarity approach? Why do you think this is? *provide a brief (1-2 paragraph) discussion based on these questions.*

#### Cosine: MAE:  0.6903937198616398   RMSE:  0.8991489107420851 
#### Pearson: MAE:  0.6917506759527966   RMSE:  0.8983386235087089 
#### Threshold: MAE:  0.6883253384684946   RMSE:  0.8936574572009663

Among these three results, the thresholded version is the best, the second one is the Cosine Similarity approach and Pearson Correlation is the worst among these three. 

Problem for Cosine Similarity is that it treats missing value as "Negative". This will greatly influence the result when you have a lot of missing values in your matrix. And the problem for Pearson Correlation is that it return a high similarity between two users with few ratings in common. And the Pearson Correlation with threshold solves this problem by setting a lower bound on number of common ratings between two users. 



# Part 2. MF (20 points)

In class, we introduced how matrix factorization works for recommendation. Now it is your term to implement it. There are different methods to obtain the latent factor matrices **P** and **Q**, like gradient descent, Alternating Least Squares (ALS), and so on. Pick one of them and implement your MF model. *You can refer to tutorials and resources online. Remember our **collaboration policy** and you need to inform us of the resources you refer to.* 

Please report MAE and RMSE of your MF model for the test set. *You are expected to get a lower MAE and RMSE compared with the results in Part 1.*

In [ ]:
a = {"k": "d"}
if a.get("k") is not None:
    print(a.get("k"))

In [188]:
# Your Code Here...
# Report Mean Absolute Error and Root Mean Squared Error for test
# gradient descent
def matrix_factorization(R, P, Q, K, steps, alpha, beta, user_dict):
    Q = Q.T
    for step in range(steps):
        for i in range(len(R)):
            for j in range(len(R[i])):
                if user_dict.get(i).get(j) is not None:
                    eij = R[i][j] - np.dot(P[i,:], Q[:,j])
                    for k in range(K):
                        P[i][k] = P[i][k] + alpha * (2 * eij * Q[k][j] - beta * P[i][k])
                        Q[k][j] = Q[k][j] + alpha * (2 * eij * P[i][k] - beta * Q[k][j])
        eR = np.dot(P,Q)
    return P, Q.T
                        


Which method did you use to obtain **P** and **Q**? What are the advantages and disadvantages of the method you pick? *provide a brief (1-2 paragraph) discussion based on these questions.*

I used gradient descent to obtain the P and Q. The advantages of gradient descent is that it is easy to understand and implement. And it can provide a better result compared with user-based collaborative fitering.  
The downside of the gradient descent is that the convergence rate is slow if you set the learning rate slow. And it might never converge if you set the learning rate too big. So it requires manual parameter tuning. Even if you have a proper learning rate, it is still a long process if you have a large dataset. 

In [191]:
N = len(train_matrix)
M = len(train_matrix[0])
K = 2

P = np.random.rand(N,K)
Q = np.random.rand(M,K)
alpha = 0.00002
beta = 0.02
steps = 1000
nP, nQ = matrix_factorization(train_matrix, P, Q, K, steps, alpha, beta, train_user_dict)
nR = np.dot(nP, nQ.T)

In [186]:
# Evaluation
def MFEval(test_user_dict, result, test_counter):
    MAE = 0
    RMSE = 0
    for user, movies in test_user_dict.items():
        for movie in movies:
            MAE += abs(movies.get(movie) - result[user][movie])
            RMSE += pow((movies.get(movie) - result[user][movie]), 2)
    MAE /= test_counter
    RMSE = math.sqrt(RMSE / test_counter)
    return MAE, RMSE

In [199]:
MF_MAE, MF_RMSE = MFEval(test_user_dict, nR, test_counter)
print("MF using Gradient Descent")
print("MAE: ", MF_MAE, " RMSE: ",  MF_RMSE)

MF using Gradient Descent
MAE:  0.687973222295372  RMSE:  0.8921783968682473


# Part 3. Extension (30 points)

Given your results in the previous parts, can you do better? For this last part you should report on your best attempt at improving MAE and RMSE. Provide code, results, plus a brief discussion on your approach. Hints: You may consider using the title or genres information, trying other algorithms, designing a hybrid system, and so on. *As in the last homework, you can do anything you like to improve MAE and RMSE.*

You will get full marks for this part if you get better results than both of your CF and MF (of course we will also judge whether what you do here is reasonable or not). Additionally, you will get 5 points as bonus if your model performs the best among the whole class.

## Method
predict value = overall average + user bias with Jaccard Similarity in movie genre + item bias



In [73]:
# Your Code Here...
# Report Mean Absolute Error and Root Mean Squared Error for test

# map each movie with a set of genres
genre_map = {}
with open ('movies_info.csv') as csv_file:
    csv_reader = csv.reader(csv_file, delimiter = '\t')
    for counter, row in enumerate(csv_reader):
        if counter > 0:
            genres = row[2].split('|')
            genre_map[counter - 1] = genres


In [112]:
def overall_average(dict):
    # calculate the overall mean
    sum = 0
    count = 0
    for user, movies in dict.items():
        for movie in movies:
            sum += dict.get(user).get(movie)
            count += 1
    return sum / count

def get_bias(dic, item):
    sum = 0
    count = 0
    for i in dic.get(item):
        sum += dic.get(item).get(i)
        count += 1
    return sum / count

def jaccard(genre_map, i, j):
    return len(list(set(genre_map.get(i)).intersection(genre_map.get(j)))) / len(list(set(genre_map.get(i)).union(genre_map.get(j))))

In [93]:
mean = overall_average(train_user_dict)
print(mean)

3.4825096064661456


In [174]:
def cal_user_bias(train_user_dict, i, genre_map, j):
    # use jaccard similarity to calculate user bias
    norm = 0
    total = 0
    count = 0
    for movie in train_user_dict.get(i):
        similar = jaccard(genre_map, movie, j)
        norm += similar
        count += 1
    norm /= count
    if norm == 0:
        return 0
    for movie in train_user_dict.get(i):
        similar = jaccard(genre_map, movie, j)
        total += train_user_dict.get(i).get(movie) * (similar / norm)
    return total / count
    

In [175]:
better_result = {}
for user, movies in test_user_dict.items():
    for movie in movies:
        user_bias = cal_user_bias(train_user_dict, user, genre_map, movie) - mean
        movie_bias = get_bias(train_movie_dict, movie) - mean
        better_result[(user, movie)] = mean + user_bias + movie_bias
        
     

0 635 0.9261844186535382 -0.31924430034369644
0 587 0.8671663291391245 -0.7110810350375742
0 339 0.9704320595070608 0.003204679248140163
0 250 1.1793176776389864 -0.14917627313281212
0 411 0.875998850170165 0.017490393533854398
0 324 1.006173494493527 0.2049903935338544
0 589 0.8826082590686108 -0.42536674932328866
0 267 0.8936149339390789 0.594413470456931
0 382 0.9907672189133776 0.20714556594764755
0 473 0.9385496872175136 -0.26432778828432735
0 639 0.946102001818184 0.017490393533854398
0 396 0.946102001818184 -0.577747701704241
0 251 1.1771389306949671 -0.2902019141584531
0 569 0.881286733280453 0.5636442396877004
0 480 1.0037641217985187 0.3905062665497274
0 203 0.9143098649389629 -0.24942689969922816
0 629 0.946102001818184 0.2374903935338546
0 663 1.0141655853497116 0.5293951554386163
0 577 0.881286733280453 0.4341570602005209
0 621 1.0199890054161442 0.2097980858415469
0 258 0.9596089072174752 -0.05393817789471722
0 519 0.8887583043377298 -0.34457857198338715
0 614 0.959948239

14 922 0.014970397916454825 0.017490393533854398
14 619 0.27942695446874577 0.45951937904110096
14 1195 -0.07975619049102622 0.0487403935338544
14 1203 0.13511568429300214 0.2362403935338544
14 91 -0.1497146850680262 0.4430223084274716
14 99 0.24818845913520082 0.8616764400454824
14 1064 -0.0593182577695095 0.47749039353385436
14 80 0.023295167379867365 0.6841570602005214
14 0 -0.3202537605239182 0.37360550144752347
14 608 0.03206727691878841 0.27868442338460087
14 767 -0.03361108714509253 -0.7444143683709075
14 707 -0.02279687218693649 0.14906934090227564
14 717 -0.04590751678395 0.3258237268671875
14 811 -0.1358347861022855 -0.2602873842439233
14 890 -0.29083093516710257 0.2648588145864861
14 1202 0.023295167379867365 -0.020971144927683927
14 491 -0.3268096902884423 -0.22325034720688652
14 1151 -0.0028717084238509827 -0.4825096064661456
14 331 -0.14798888107754626 0.36285121827612254
14 945 -0.07975619049102622 -0.28685743255310214
14 1183 -0.025805423264011917 0.284157060200521
14 4

18 136 -0.5708387903796353 -0.597894221850761
18 0 -0.5821326586395106 0.37360550144752347
18 431 -0.6295941063005768 0.13513745235738384
18 705 -0.7105515275276382 0.0545274305708916
18 741 -0.7105515275276382 -1.1491762731328121
18 188 -0.4570030941651315 0.08891896496242602
18 466 -0.6402468379220103 -0.5450096064661456
18 396 -0.6007214322869117 -0.577747701704241
18 358 -0.7476796274350366 -0.30859656298788485
18 731 -0.5784012792738089 0.4460618221052828
18 427 -0.6007214322869117 -0.5859578823282146
18 289 -0.598845793797881 0.5583067200644662
18 409 -0.7370684447570426 -0.7456675012029876
18 447 -0.6702597880651444 -0.5658429397994791
18 143 -0.6297933086794258 0.05595193199539272
18 445 -0.7370684447570426 -1.0088253959398297
18 171 -0.5887299189923767 0.2503148210147703
18 572 -0.7370684447570426 -0.3916005155570548
18 80 -0.5708387903796353 0.6841570602005214
18 711 -0.7341382709612603 -0.2825096064661454
18 30 -0.7370684447570426 -1.0325096064661454
18 154 -0.73706844475704

22 301 0.029819160657144295 0.7079665840100451
22 217 0.029819160657144295 0.7026755787190395
22 130 0.029819160657144295 0.017490393533854398
22 332 0.01074665557624499 0.5174903935338544
23 869 -0.02835507740524834 0.21749039353385458
23 499 -0.00824720847493321 0.06294493898839981
23 918 -0.005711382147632005 0.1760269788997082
23 25 0.07845650183147734 0.6548949736865262
23 180 0.058297181482370064 0.5748674427141824
23 558 -0.07644440555636534 -0.6295684299955573
23 1145 -0.010942393610610779 0.05915706020052092
23 803 -0.07644440555636534 -0.30603901823085167
23 99 0.16204666068847384 0.8616764400454824
23 1021 0.02143772999266913 -0.07074490058379279
23 1023 0.07315590565656649 0.017490393533854398
23 1013 0.006266041178987436 -0.7325096064661456
23 828 -0.02487601867988598 0.40121132376641233
23 964 0.0013174279226855212 0.3091570602005209
23 305 -0.05607818169916756 0.19021766626112724
23 974 -0.008549395394727988 -0.2411302961213182
23 1113 0.05357919957017421 0.4237403935338

31 361 0.4616964021175458 0.3595956566917491
31 497 0.33113868749710873 -0.03806516202170096
31 92 0.4249890825742333 0.512705704538639
31 437 0.4695661400742783 -0.4325096064661458
31 608 0.4662458840170167 0.27868442338460087
31 284 0.48426090236044006 0.24825962430308524
31 605 0.33113868749710873 -0.8491762731328123
31 7 0.40661895263653136 0.25218427108487473
31 0 0.41028771346684945 0.37360550144752347
31 84 0.4119241747238722 0.24825962430308524
31 387 0.4119241747238722 0.2821962758867955
31 139 0.3155359961397175 0.3896708446616741
31 480 0.5313792824227463 0.3905062665497274
31 247 0.4361556698633624 -0.36486254764261616
31 113 0.41981203931099786 0.530823726867188
31 523 0.49198361410037705 0.02692435579800545
31 492 0.9230105633852337 -0.09541283227259711
31 150 0.42856911073502246 -0.07074490058379279
31 377 0.4616964021175458 -0.06432778828432761
32 172 -0.15292006184699058 0.34755575301097874
32 811 -0.47191306015531476 -0.2602873842439233
32 558 -0.8314170913568937 -0.6

40 610 0.16077049401833587 -0.4055865295430685
40 205 0.17053402174560928 -0.9682238921804314
40 416 0.2736177264174775 0.3723291032112739
40 520 0.17214320971206742 -0.08250960646614569
40 415 0.3910120345548469 -0.3354507829367339
40 670 0.5099715965413716 -0.18839195940732223
40 111 0.16151226184856693 0.08891896496242602
40 609 0.08772377090065442 0.053204679248139986
40 99 0.5494092692271892 0.8616764400454824
40 392 0.16077049401833587 -0.8575096064661456
40 164 0.1397452235656491 -0.3916005155570548
40 767 0.2411724295609532 -0.7444143683709075
40 156 0.5303975846832105 0.5395492170632665
40 262 0.2675960124607677 0.3950414139420175
40 422 0.2391194827430949 0.1319482248591557
40 140 0.118986427400118 -0.677631557685658
40 701 0.16077049401833587 -0.3325096064661457
40 185 0.3527518886357961 -0.19679532075186001
40 71 0.1493775226884364 -0.5866762731328121
40 622 0.20348085733966004 0.1737403935338544
40 90 0.16805053031765294 -0.12044064094890405
40 253 0.17214320971206742 0.34

47 646 -1.0157447673179814 -0.0010281249846642027
47 290 -0.9401417955320341 0.6147126157560767
47 589 -0.9270286596808517 -0.42536674932328866
47 819 -1.048383837985901 0.21160804059267813
47 904 -0.8368785385049806 0.2674903935338544
47 702 -0.7016380647536011 0.04380618300753847
47 642 -0.8507449005837939 0.46343633947980045
47 555 -0.6560950110140045 -0.020971144927683927
47 178 -1.082851681722703 -0.11294438907484139
47 994 -1.0370002634884479 -0.21935171172930357
47 745 -0.8434924594246707 -0.3825096064661455
47 1106 -0.5078909365454605 0.3508237268671879
48 697 0.19812623168414456 -0.06946612820527598
48 620 0.10992221386036549 0.4508237268671875
48 664 0.6291655204374078 0.017490393533854398
48 415 0.1817498794579686 -0.3354507829367339
48 584 0.025710896435207076 -0.5579813045793531
48 660 0.22288192113334038 -0.30069142464796395
48 217 0.5923093798572663 0.7026755787190395
48 526 0.19812623168414456 -0.5102873842439233
48 140 0.07276465513723362 -0.677631557685658
48 125 -0.0

53 397 -0.3430784691780828 -1.44679532075186
53 550 -0.16030880507505163 -0.12139849535503444
53 392 0.004788192526161694 -0.8575096064661456
53 729 -0.2902598282055462 -0.2950096064661456
53 234 0.11139039778176363 0.1303936193403059
53 580 0.004788192526161694 0.16454921706326608
53 406 -0.10676129689209013 -0.69679532075186
53 204 0.19436405198989926 -0.41584293979947873
53 173 -0.01978411977208605 0.2821962758867955
53 523 0.012948787806809747 0.02692435579800545
53 692 -0.09935159137353278 -0.16672013278193498
53 723 -0.011199781866195302 0.24722012326358422
53 444 -0.3239837464384099 -0.2325096064661456
53 637 0.03068700412424974 -0.026627253524968975
53 430 0.004788192526161694 -0.2719832906766717
53 518 -0.1759608597813962 0.2362403935338544
53 534 0.04172034385322698 -0.1811397434524471
53 351 -0.3051614941234484 0.48891896496242593
53 401 -0.182764354951519 -0.21057978190474191
53 46 -0.08792623194800386 -0.6044608259783408
53 638 -0.3950663367056588 -0.17481729877383811
53 1

59 617 0.05497249777095892 -0.9269540509105902
59 443 0.06824505485731835 -0.1031992616385593
59 171 0.053356001102747186 0.2503148210147703
59 882 0.04124507380695208 -0.21935171172930357
59 763 0.01311064577824217 0.41604111817153555
59 685 0.028788916655472008 -0.1031992616385593
59 791 0.09863132188469992 -0.1311582551147943
59 766 0.06571813933206716 -0.0010281249846642027
59 112 0.03227815837584913 -1.2602873842439233
59 762 -0.03969543289120825 0.5730459490894098
59 806 0.06324766862117981 -0.3246148696240403
59 736 0.0958419902211376 -0.4454725694291084
59 488 -0.0026954294266174905 -0.10015666528967504
59 672 0.028788916655472008 -0.5825096064661457
59 99 0.18703558639510964 0.8616764400454824
59 743 -0.0367386260416982 -0.7602873842439233
59 273 0.10731411677839642 0.18415706020052092
59 761 0.08555055891189545 0.0603475363909971
59 951 0.14854290057852637 0.3299903935338544
59 507 0.11810359346930666 0.11008298612644696
59 524 0.06571813933206716 -0.21327883723537644
59 71 -

63 299 -0.3118033044947883 0.6130786288279721
63 623 -0.49707624885506796 -0.05393817789471722
63 408 -0.3070214280805068 0.3143653935338544
63 864 -0.5659359269460373 -0.4825096064661456
63 376 -0.4439447879530767 -0.5182238921804312
63 103 -0.49707624885506796 0.21924477949876664
63 1177 -0.3943493551228827 -0.08250960646614569
63 792 -0.2938476120054738 -0.9491762731328124
63 1094 -0.4316498376425253 0.3263139229456189
63 233 -0.43224485530463364 0.3239420064370804
63 997 -0.35369227611530807 -0.07074490058379279
63 18 -0.5222030180773625 0.1319482248591557
63 662 -0.4563236497813272 0.2674903935338544
63 595 -0.5732305640141049 -0.7855399094964488
63 160 -0.3738998203618129 0.46810767748447146
63 938 -0.5058238409895783 -0.20978233373887267
63 372 -0.41714117211595836 -0.7456675012029876
63 1017 -0.46315870200949316 0.02860150464496547
63 818 -0.48795118994655917 0.30127417731763817
63 969 -0.4381690711549653 -0.1075096064661456
63 979 -0.4752185121679906 -0.2411302961213182
63 908

63 643 -0.46289524476401933 -0.7982990801503562
63 561 -0.5111784452690049 -0.24441436837090746
63 679 -0.44914673682887507 -1.038065162021701
63 501 -0.5311252647851674 -0.3991762731328121
63 912 -0.4202563446196521 -0.411081035037574
63 614 -0.4236431910887033 0.08645591077523385
63 1110 -0.2976274128756349 -0.276627253524969
64 25 0.7045871677274032 0.6548949736865262
64 113 1.013658975909332 0.530823726867188
64 575 0.8990693409022747 0.5280724041158646
64 723 1.3204101015630507 0.24722012326358422
64 817 0.9858781701408108 -0.2602873842439233
64 816 0.8644689510387278 0.16454921706326608
64 799 0.9913859758631713 -0.3030224269789663
64 631 1.1107107325169054 0.15385402989749064
64 668 1.0250280819760644 0.19140343701211515
64 415 0.8429805896122855 -0.3354507829367339
64 92 0.903191668579391 0.512705704538639
64 172 0.9790288550723161 0.34755575301097874
65 287 0.7701785655768663 0.617490393533854
65 345 0.25082372686718823 0.49310014963141535
65 689 0.6768441662527827 0.244763120

78 608 -0.18272774295726402 0.27868442338460087
78 907 -0.6651497511316058 0.24198018945222177
78 663 0.10082372686718744 0.5293951554386163
78 1113 -0.36297911096061064 0.4237403935338544
78 335 -0.47080328250635395 0.3699494099272971
78 848 -0.7657575158268144 0.12999039353385422
78 499 -0.6534079181042451 0.06294493898839981
78 1099 -0.5810287965537859 0.13287500891846982
79 217 0.5047862193051822 0.7026755787190395
79 11 0.03067221171567258 -0.1432238921804312
79 335 0.0020870955663236934 0.3699494099272971
79 109 0.3825656884193962 0.14825962430308515
79 288 0.1495128654439677 0.6769106833889267
79 373 0.16531440118569574 -0.5477269977704933
79 334 0.061169464528842976 0.20170091984964378
79 831 0.3981817931965894 0.4237403935338544
79 139 0.16278097469617858 0.3896708446616741
79 270 0.20138848900751372 -0.4522065761631153
79 221 0.16531440118569574 0.4852323290177254
79 332 0.5551859838610249 0.5174903935338544
79 99 0.5047862193051822 0.8616764400454824
79 262 -0.05637324282978

84 225 -0.10780941415707224 0.6603475363909976
84 122 -0.031118640268120945 -0.21327883723537644
84 818 -0.07741281500639507 0.30127417731763817
84 147 -0.04011869627637221 -0.07510219905873816
84 287 -0.42117859298202465 0.617490393533854
84 378 -0.7156538406144448 0.3508237268671879
84 548 -0.11766013185454804 0.13653801258147347
84 80 0.01812292565982654 0.6841570602005214
84 772 -0.3230183895949792 0.21898293084728726
84 437 -0.12916778453089206 -0.4325096064661458
84 352 -0.7620786429016735 0.17266280732695805
84 110 -0.13957445508807886 -0.13367239716381984
84 39 -0.22639456188448248 0.0336194257919189
84 73 -0.4591061296651522 -0.11887324282978184
84 171 -0.27459467552994665 0.2503148210147703
84 159 -0.61549985375453 -0.0645991587049517
84 460 -0.023796622411472068 0.1603475363909972
84 176 -0.16062050735393152 0.044763120806581735
84 90 -0.07463306462201835 -0.12044064094890405
84 593 -0.058550438538320204 0.5174903935338544
84 630 -0.04011869627637221 0.1674903935338543
84 44

95 1078 0.6149764476463564 -0.14917627313281212
95 724 0.46588788932903746 0.38112675717021816
95 347 0.788605458777035 0.11749039353385449
95 556 0.28546355361479847 -0.6491762731328121
95 571 0.36227337714983987 0.41404211767178545
95 283 0.806600602968667 -0.01192137117202785
95 1012 0.15280764107629974 -0.3158429397994791
95 38 0.71695015165283 0.5331153935338544
95 346 -0.13683423852450805 0.365975242018703
95 1103 0.788605458777035 0.05595193199539272
95 633 0.5514888170706356 0.6532046792481405
95 1068 -0.033527191923566235 -0.889917013873553
95 1 -0.0648042933660209 -0.17871213811171538
95 935 -0.13878663646659817 0.24883367711594406
95 733 0.15280764107629974 -0.015842939799478817
95 968 0.024786498473384988 -0.7995827771978528
95 847 0.5500974110416013 -0.3955530847270152
95 1094 0.2237525498972679 0.3263139229456189
95 345 0.28449878123189976 0.49310014963141535
95 849 -0.15036812698482294 0.22582372686718788
95 944 0.28449878123189976 0.07304594908940976
95 800 0.2844987812

103 881 -0.3419055796205126 -0.2889612193693716
103 757 -0.3419055796205126 -1.215842939799479
103 1010 -0.4797497584840009 -0.4179934774338876
103 834 -0.4073455438074167 -0.03806516202170096
103 504 -0.30104346409812965 -0.2920334159899549
103 1024 -0.3419055796205126 -0.32341869737523643
103 397 -0.5141228474923811 -1.44679532075186
103 934 -0.5161668912847444 -0.5413331358779101
103 1112 -0.41305483386732034 -0.25523687919341853
103 1175 -0.30199970570963286 0.15851603455949537
103 381 -0.6377871635608883 -0.4200096064661456
103 676 -0.2877044116609513 -0.1682238921804311
103 812 -0.35021737462447833 -0.569466128205276
103 763 -0.4324059527831605 0.41604111817153555
103 1187 -0.59411919175994 0.3174903935338542
103 879 -0.3267570697612294 -0.429878027518777
103 1059 -0.5373846762595553 -0.047726997770493274
103 1105 -0.5869849762270127 0.030310906354367173
103 401 -0.4892319301464667 -0.21057978190474191
103 922 -0.5204673448355597 0.017490393533854398
103 24 -0.37821464550015005 -

109 29 -0.06644462451668698 0.08611784451424676
109 56 -0.39597723601131296 -0.5575096064661458
109 90 -0.2706927621048463 -0.12044064094890405
109 99 -0.1279770285624613 0.8616764400454824
109 57 -0.09996139274499516 -0.8764490004055396
109 168 -0.1452197943988871 -0.8825096064661455
109 209 -0.1321538752408462 0.0799903935338544
109 111 -0.36652645011316043 0.08891896496242602
109 10 -0.1452197943988871 -0.8461459701025094
109 146 -0.26069651861343246 -0.7083160580790486
109 172 -0.44218702582098324 0.34755575301097874
109 202 -0.09957007060338618 0.3587602348036958
109 145 -0.3941735612490507 0.282591064674794
109 25 -0.30281051767614864 0.6548949736865262
109 152 -0.07621792127272364 0.04380618300753847
109 188 0.060499944113594495 0.08891896496242602
110 801 0.45444895124867557 -0.34614597010250936
110 0 0.48296963556245753 0.37360550144752347
110 597 0.4953640521244096 -1.6115418645306618
110 857 0.5773491120863379 -0.11408855383456684
110 790 0.5220822829157572 0.224386945257992

121 589 -0.7460776299351548 -0.42536674932328866
121 221 -0.5939790357660497 0.4852323290177254
121 49 -0.5649112054212821 0.13112675717021816
121 951 -0.44425474464424175 0.3299903935338544
121 106 -0.7610973215217651 -0.5658429397994791
121 486 -0.3424637481815984 0.3823552583987193
121 99 -0.30128817154083753 0.8616764400454824
121 27 -0.4931793362523238 0.5174903935338544
121 327 -0.44425474464424175 0.45191662304205105
121 107 -0.492435688831514 -0.1253667493232884
121 267 -0.2894736978698478 0.594413470456931
121 1003 -0.7540023427985187 -0.5325096064661454
121 230 -0.4760760483161084 0.5974903935338545
121 844 -0.6967217441524887 -0.15692821111730826
121 999 -0.36669820637270734 0.18415706020052092
121 72 -0.492435688831514 -0.017393327396378222
121 1080 -0.36254385274532064 0.0392295239686371
121 149 -0.492435688831514 -0.1555865295430685
121 317 -0.4931793362523238 0.5651094411529018
121 113 -0.48714577776771195 0.530823726867188
121 116 -0.5327122913971443 0.40824669605486275

128 775 0.04381718264088352 -0.215842939799479
128 272 0.09738005945538797 -0.05393817789471722
128 428 0.05666002036246809 0.003976880020340801
128 18 0.14720903362764215 0.1319482248591557
128 875 0.11410907598597664 0.23543911148257246
128 512 0.06392659320160643 0.4674903935338546
128 676 0.07238219736040552 -0.1682238921804311
128 871 0.0605855553409711 0.3174903935338542
128 375 -0.05401996945709975 0.3214119621613052
128 195 -0.051742264538585214 0.017490393533854398
128 253 -0.05759572507707844 0.34101980529856046
128 792 0.1463217894962785 -0.9491762731328124
129 1033 -0.3272437503067769 0.3081880679524591
129 645 -0.27905944217260714 -0.03514118541351419
129 919 0.0932581680395339 0.13287500891846982
129 662 -0.03271482923694924 0.2674903935338544
129 876 -0.18767886922163068 -0.1682238921804311
129 71 -0.18560364700874343 -0.5866762731328121
129 135 -0.36032816965439673 0.3508237268671879
129 905 -0.3165226120872173 -0.3361681430515113
129 1046 0.12097974171998205 0.13287500

139 301 0.6180137254134901 0.7079665840100451
139 1127 0.351881326299873 -0.13468351950962365
139 1078 0.5099538885880235 -0.14917627313281212
139 224 0.5198638112553717 0.49788255039659957
139 1101 0.6459590471823637 -0.032509606466145424
139 99 0.6180137254134901 0.8616764400454824
139 225 0.20825202964104816 0.6603475363909976
139 175 0.40322629970108803 0.6530836138728371
139 290 -0.1807687940870344 0.6147126157560767
139 306 0.6180137254134901 0.6187562163186642
139 1148 0.5381800487062676 0.4405673166107773
139 156 0.6858752732589406 0.5395492170632665
139 313 0.3477424943741898 0.3424903935338546
140 1131 -1.792944066176506 -0.05393817789471722
140 169 -1.9543144214604908 -0.5871607692568435
140 1176 -1.2107975821790373 0.2952681713116321
140 781 -1.6753879447450775 0.4103475363909972
140 1058 -1.3668414207746182 0.029118300510598605
140 138 -1.3839689891343623 -1.1346835195096237
140 523 -1.1899063728455044 0.02692435579800545
140 633 -1.545288662074318 0.6532046792481405
140 8

145 646 -0.6569493214797988 -0.0010281249846642027
145 162 -0.8979687406306871 -1.7682238921804314
145 177 -0.1278345709718285 0.14670387667992202
145 472 -0.9780927441157039 -0.07267354089237532
145 799 -0.7293917923176263 -0.3030224269789663
145 888 -1.0303060850133607 -0.3731346064661456
145 203 -1.0548607719038663 -0.24942689969922816
145 553 -0.5251686694204127 -0.7132788372353764
145 121 -0.49252510664499205 -1.4825096064661456
145 431 -1.0256732119576228 0.13513745235738384
145 352 -0.6755452647323787 0.17266280732695805
145 483 -0.8434313851094424 -0.9172922151617979
145 785 -0.3531820669305965 -0.027964151920691016
145 405 -1.0548607719038663 -0.9825096064661456
145 0 -0.7143964091283372 0.37360550144752347
145 554 -0.5251686694204127 -1.0088253959398297
145 469 -0.8794563131252628 -0.21935171172930357
145 145 -1.0548607719038663 0.282591064674794
145 608 -0.17957666004682693 0.27868442338460087
145 636 -0.8437818155326875 -1.3713984953550344
145 403 -1.0227215417701063 -0.569

155 233 1.131707492477179 0.3239420064370804
155 166 1.3023005201161335 0.0799903935338544
155 363 1.1104882053500478 -0.69679532075186
155 205 1.057832691822362 -0.9682238921804314
155 210 1.1093271282277333 -0.2756130547420077
155 0 1.0854515585823976 0.37360550144752347
155 23 1.1582089564081062 0.4470678583225869
155 15 1.1107378107241819 0.402105778149239
155 14 1.1093271282277333 0.17055161802365015
155 80 1.164348951255921 0.6841570602005214
155 280 1.0887928997110525 0.4032046792481401
155 17 1.2106971137676017 0.45442733047079154
155 49 1.0449165538714076 0.13112675717021816
156 462 0.5174903935338553 0.34101980529856046
156 335 0.01475815309669537 0.3699494099272971
156 463 0.4021057781492394 -0.2973244212809605
156 188 0.11089698694044836 0.08891896496242602
156 927 -0.019144613250407172 0.23807862882797215
156 156 -0.10515111590010795 0.5395492170632665
156 299 -0.2488647466530609 0.6130786288279721
156 303 0.11887288201311641 0.365975242018703
157 145 -0.18813731209385098 

162 198 0.07889390230578286 -1.1491762731328121
162 109 0.46257689127271107 0.14825962430308515
162 122 0.33613446133046354 -0.21327883723537644
162 45 0.17787917405446985 -0.6211234678522843
162 68 0.663719368031956 0.031003907047367996
162 175 0.3979344584356479 0.6530836138728371
162 50 0.3129449389883985 -0.015842939799478817
162 259 0.4658108069705471 -0.05393817789471722
162 78 0.9283708516584066 -0.14917627313281212
162 184 0.07889390230578286 -0.5625096064661457
162 37 0.6185350279118205 0.4356190485046145
162 180 0.42857400099892473 0.5748674427141824
162 203 0.26713007434378166 -0.24942689969922816
162 172 0.15654364797172304 0.34755575301097874
163 827 -0.8666559479295599 0.4644600905035512
163 714 -0.32461486962403985 0.31160804059267777
163 289 -0.33404238784900553 0.5583067200644662
163 206 -0.32461486962403985 -0.3671249910815302
163 575 -0.2179002827629284 0.5280724041158646
163 80 -0.21867981923210378 0.6841570602005214
163 279 -0.5559107692568439 0.605402481445942
163

169 487 0.7553159164272523 0.2280167093233283
169 369 0.6685593220827619 -0.42536674932328866
169 575 0.5932961130073031 0.5280724041158646
169 449 0.5900576622400853 0.2097980858415469
169 507 0.7237819087613242 0.11008298612644696
169 680 0.6401348991446967 -0.8509306590977244
169 125 0.6489330201503356 -0.14084293979947882
169 650 0.701691513514795 -0.05393817789471722
169 233 0.7463773133075606 0.3239420064370804
169 733 0.7149744375630864 -0.015842939799478817
169 365 0.6970350331896431 -0.7083160580790486
169 334 0.5989402869240492 0.20170091984964378
169 731 0.6922499134703277 0.4460618221052828
169 431 0.6333395552703638 0.13513745235738384
169 97 0.6685593220827619 -0.09120525864005868
169 637 0.7151848776921814 -0.026627253524968975
169 241 0.5659248983261889 0.14906934090227564
169 302 0.634741544791384 0.4174903935338543
169 174 0.5349741501788672 -0.13922602437659348
169 711 0.5873390547678468 -0.2825096064661454
169 478 0.6970350331896431 -0.3325096064661457
169 625 0.628

179 645 -0.24173771136470767 -0.03514118541351419
179 529 0.3911800309285973 0.5619348379782987
179 228 0.3378768442887665 0.6479251761425497
179 229 0.44560335887403335 0.5820065225661124
179 106 -0.419559966178376 -0.5658429397994791
179 411 0.30794876584442576 0.017490393533854398
179 174 0.1215660849457918 -0.13922602437659348
179 64 0.3378768442887665 0.2508237268671878
179 827 0.04834639884241465 0.4644600905035512
179 202 0.14151565124621035 0.3587602348036958
179 415 0.30794876584442576 -0.3354507829367339
179 83 0.3203463141710383 -0.14917627313281212
179 223 0.28583181517245126 0.46987134591480695
179 92 0.14177723261083752 0.512705704538639
179 813 0.3085022490618412 0.14606182210528296
179 13 0.2013475799848794 -0.5325096064661454
179 270 0.31123428405193687 -0.4522065761631153
179 612 -0.419559966178376 0.2774903935338542
179 406 0.28583181517245126 -0.69679532075186
179 748 0.41125482771399335 0.22119409723755812
179 302 -0.24123744415578052 0.4174903935338543
179 25 0.31

192 957 0.06727626077154314 -0.3158429397994791
192 972 -0.09524021903702584 0.4460618221052828
192 1207 -0.04451032536259625 0.15749039353385452
192 1135 -0.0671034144316387 0.3612403935338544
192 763 -0.06886655669904407 0.41604111817153555
192 1127 -0.025387236928286594 -0.13468351950962365
192 1063 -0.049611514833601156 0.5820065225661124
192 305 -0.055366637287256104 0.19021766626112724
192 1083 -0.049201487454546555 0.2821962758867955
192 970 -0.01851679666949213 -0.10155722551376467
192 859 -0.06533929808244654 0.4312834969821302
192 1010 -0.0027816105718652118 -0.4179934774338876
192 299 -0.003663690752273574 0.6130786288279721
192 1142 -0.02533286019446601 0.2774903935338542
193 1068 0.3249727527403512 -0.889917013873553
193 552 0.33918157000444227 -0.30603901823085167
193 1064 0.3010707548122231 0.47749039353385436
193 871 0.10172725758723011 0.3174903935338542
193 837 0.28304063276830194 -0.6325096064661455
193 1044 0.2958671530426895 -0.19679532075186001
193 426 0.246394555

196 706 -0.629195227484415 -0.17695405091059024
196 66 -0.8355805884448162 0.2187890948325557
196 213 -0.7743759724091985 -0.25523687919341853
196 166 -0.5334565058348772 0.0799903935338544
196 701 -0.8355805884448162 -0.3325096064661457
196 509 -0.8236967383193399 -0.2825096064661454
196 466 -0.5648179302447813 -0.5450096064661456
196 351 -0.49384496726238325 0.48891896496242593
196 454 -0.4820380195762626 0.2374903935338546
197 411 0.2641570602005219 0.017490393533854398
197 908 -0.4127049314741509 -0.5916005155570545
197 290 -0.2325096064661456 0.6147126157560767
197 718 -0.09388401404908464 -0.2825096064661454
197 393 -0.17549206260649663 -0.051954050910590244
197 142 -0.20343983902428464 0.017490393533854398
197 352 -0.12618776738568593 0.17266280732695805
197 268 0.9273264591076251 0.42446713771990074
198 306 -0.2155353700635998 0.6187562163186642
198 946 -0.37180312816546346 -0.06028738424392355
198 208 -0.6177636513142857 -0.7875943522288575
198 813 -0.27575302165891946 0.14606

200 806 0.4303891995624318 -0.3246148696240403
200 951 0.47357046523985646 0.3299903935338544
200 220 0.4467999415512325 0.8508237268671874
201 927 -0.427852440446602 0.23807862882797215
201 1145 -0.5223892591919643 0.05915706020052092
201 1175 -0.3580690213348925 0.15851603455949537
201 1058 -0.4092401211434007 0.029118300510598605
201 598 -0.4549080483136687 -0.3942743123484984
201 293 -0.5276183982581135 0.6216570602005214
201 119 -0.4861378509540004 -0.5658429397994791
201 156 -0.3029395803694648 0.5395492170632665
201 301 -0.26308865656373825 0.7079665840100451
201 921 -0.4549080483136687 -0.8671249910815302
201 136 -0.5551816562476097 -0.597894221850761
201 1121 -0.32204921377826246 0.2424903935338545
201 50 -0.4669855489982053 -0.015842939799478817
201 697 -0.42376643706723893 -0.06946612820527598
201 1113 -0.4064930742753843 0.4237403935338544
201 554 -0.36179965446034945 -1.0088253959398297
201 182 -1.074225582797506 -0.13768202025924925
201 512 -0.3302135027686717 0.467490393

211 668 -0.32934482381094554 0.19140343701211515
211 855 -0.24950830720239336 -0.527964151920691
211 1004 -0.17321566766027408 0.09720053846139054
211 898 -0.4463445642648849 -0.4825096064661456
211 1083 0.2280786288279706 0.2821962758867955
211 732 -0.3182030853180482 -0.12250960646614573
211 994 -0.4164716569195632 -0.21935171172930357
211 880 -0.05873608791506646 -1.1825096064661458
211 450 -0.32220220455162396 0.07154444758790834
211 248 -0.4549386267622375 0.38983081906576933
211 835 -0.20924164654289124 -0.6700096064661456
211 1096 -0.026496148585902102 0.4374903935338543
211 766 -0.43904615679087966 -0.0010281249846642027
211 98 -0.180592493412862 -0.3552368791934182
211 794 -0.32734728947060354 -0.5491762731328125
211 757 -0.31649398146614516 -1.215842939799479
211 808 -0.38474844867876357 -0.4825096064661456
211 1098 -0.3027679440564528 0.2424903935338545
211 916 -0.07051838616501271 0.19749039353385456
211 703 -0.09573896787517855 0.08505796110142194
211 42 -0.097283257572218

213 372 0.013406510807668415 -0.7456675012029876
213 367 0.3180302429442814 -0.8509306590977244
213 575 0.24986340338980817 0.5280724041158646
213 585 0.289094282311114 -0.5554262731328121
213 174 0.12920124939360278 -0.13922602437659348
213 706 0.0419723499262159 -0.17695405091059024
213 469 0.08908971758897488 -0.21935171172930357
213 28 0.19905195509541374 -0.2325096064661456
213 637 0.13548344795302913 -0.026627253524968975
213 244 -0.20705668480245398 -1.204731828688368
213 699 0.11831743476872614 -0.4055865295430685
213 584 0.10902480289860739 -0.5579813045793531
213 136 0.289094282311114 -0.597894221850761
213 790 0.19049955491511028 0.2243869452579923
213 30 -0.32098544289737463 -1.0325096064661454
213 360 0.24632198096334923 0.3143653935338544
214 119 -0.1948065902248466 -0.5658429397994791
214 7 0.13888994206658456 0.25218427108487473
214 139 0.07855909582393039 0.3896708446616741
214 113 0.16983876167773015 0.530823726867188
214 63 0.28171803581027755 -0.6729857969423363
214

222 866 0.8269592157047572 -0.1736860770543811
222 1071 0.7586727834709586 0.24476312080658147
222 1025 0.651801341709159 0.07681242743215932
222 1122 0.715446601818702 0.2674903935338544
222 531 0.45844634479908297 -0.2682238921804312
222 402 0.5588669862042885 -0.05393817789471722
222 888 0.6636947368141111 -0.3731346064661456
222 315 0.6141636287256973 0.2892295239686371
222 991 0.6711399331661405 -0.19679532075186001
222 479 0.6127757823286211 -0.1794793034358424
222 1155 0.6831632568566666 -0.26511830211831944
222 1001 0.6614190542555831 -0.3212192838855006
222 1029 0.6294590972767202 0.3214119621613052
222 1090 0.5779375186895641 0.42225229829575905
222 1133 0.7367695073489275 -0.215842939799479
222 824 0.5080067265159438 0.6802810912082733
222 25 0.6431568063289794 0.6548949736865262
222 1081 0.6294590972767202 -0.3575096064661456
222 1069 0.6320982143675029 -0.42190354586008505
222 593 0.483226678423176 0.5174903935338544
222 394 0.5896404656839258 0.08159295763641872
222 1145 

225 149 0.2349279267608657 -0.1555865295430685
225 1093 0.16011612013023635 0.11749039353385449
225 1009 0.11534340066259352 -0.01192137117202785
225 626 0.07337214684679738 0.09066112524117154
225 1187 0.1326104140851645 0.3174903935338542
225 962 0.18712038011583143 -0.42190354586008505
225 725 0.14923086401671348 -0.8354507829367339
225 1135 0.21362341904735693 0.3612403935338544
225 224 0.2549278952796157 0.49788255039659957
225 442 0.19082691551928344 0.3508237268671879
225 901 0.11688662252339377 -0.16672013278193498
225 17 0.20431005462687946 0.45442733047079154
225 217 0.42397800303501976 0.7026755787190395
225 984 0.06065044413321541 -0.3543044782610174
225 775 0.2502486860093951 -0.215842939799479
225 1168 0.20892883811511087 0.0799903935338544
225 1159 0.10845840307395616 0.05082372686718761
225 596 0.11202420580935257 0.1603475363909972
225 389 0.07337214684679738 -0.5641422595273702
225 489 0.09891500732922598 -0.017393327396378222
225 946 0.21877732307990527 -0.0602873842

235 698 0.2234400502844256 0.3595956566917491
235 33 0.1770966512241352 0.22337274647503103
235 336 0.22268103375240855 -0.17816178037918906
235 510 0.3099114461654353 0.46987134591480695
235 314 0.09335096628479667 0.6753851303759597
235 15 0.16663489982121726 0.402105778149239
235 412 0.09335096628479667 0.0336194257919189
235 969 0.15400389033995499 -0.1075096064661456
235 593 0.40870378265519314 0.5174903935338544
235 906 0.3315013340667661 0.15851603455949537
235 948 0.09335096628479667 -0.049176273132812476
235 824 0.03664641796472479 0.6802810912082733
235 264 0.3076208529779829 0.4804533564968172
235 193 0.19396739375186423 0.6424903935338544
236 480 0.18596865440342025 0.3905062665497274
236 1116 0.13975008999253946 0.2248074667045863
236 752 0.13975008999253946 0.23880186894369038
236 92 0.056079605152110634 0.512705704538639
236 823 0.16485255140771438 0.12860150464496556
236 299 0.02945211602189346 0.6130786288279721
236 1064 0.0631349928369902 0.47749039353385436
236 529 0

249 930 -0.38271773441006296 0.0002490142235096471
249 854 -0.38271773441006296 -0.808091001814983
249 1011 -0.37951822194936646 -0.2773814013379403
249 405 -0.4151843855508064 -0.9825096064661456
249 1022 -0.28016007315426483 0.10572568765150159
249 518 -0.2219831077764698 0.2362403935338544
249 875 -0.17917509449787783 0.23543911148257246
249 435 -0.29132248907386193 -0.5734186973752364
249 593 -0.2675048341849995 0.5174903935338544
249 315 -0.37951822194936646 0.2892295239686371
249 160 -0.37951822194936646 0.46810767748447146
249 401 -0.4010004898834678 -0.21057978190474191
249 59 -0.2815444326426393 -0.8528799768365158
249 393 -0.39666185863930314 -0.051954050910590244
249 407 -0.34545433064033926 -0.3825096064661455
249 100 -0.4010004898834678 -0.8158429397994791
249 699 -0.41123358437045754 -0.4055865295430685
249 939 -0.38271773441006296 -0.2425096064661454
249 729 -0.4387065955990601 -0.2950096064661456
249 136 -0.4387065955990601 -0.597894221850761
249 307 -0.2078100859432443

250 697 0.5763588950629059 -0.06946612820527598
250 575 0.5843834827246419 0.5280724041158646
250 668 0.5753246527226596 0.19140343701211515
250 417 0.5903452235392517 0.11749039353385449
250 20 0.5553320193222224 -0.3230893166110733
250 305 0.5638878745488616 0.19021766626112724
250 528 0.6534414073290988 0.610082986126447
250 412 0.6414737378931172 0.0336194257919189
250 780 0.7922156682591326 -0.0025096064661456197
250 224 0.7150771205323454 0.49788255039659957
250 836 0.5992031747246691 -0.019546643503182803
250 362 0.7511884241684261 0.3962782723217333
250 866 0.7747251202219565 -0.1736860770543811
250 845 0.6414737378931172 0.032641908685369536
250 441 0.6414737378931172 -0.4236860770543811
250 577 0.6543281073427791 0.4341570602005209
250 517 0.6414737378931172 0.5174903935338544
250 237 0.6835884213278143 0.21193483797829904
250 332 0.7387697669020041 0.5174903935338544
250 688 0.7087472241349464 -0.14917627313281212
250 238 0.7511884241684261 0.5730459490894098
250 749 0.57635

261 263 -0.1734624626864245 0.3299903935338544
261 541 -0.15819556888959063 -0.6825096064661458
261 244 -0.1729217446346607 -1.204731828688368
261 744 -0.20565587478926695 -0.2902019141584531
261 550 -0.2460786482795756 -0.12139849535503444
261 845 -0.15598973698120266 0.032641908685369536
261 326 -0.3917655048327515 0.5545274305708916
261 196 -0.3096787217613586 -0.17481729877383811
261 923 -0.20535523235132258 -0.3100958133626972
261 584 -0.2717670933581475 -0.5579813045793531
261 978 -0.2799472470037192 -0.10996058685830246
261 262 -0.15819556888959063 0.3950414139420175
261 228 -0.13623301072146754 0.6479251761425497
261 647 -0.18839195940732267 0.39984333471032496
261 698 -0.0687800510641634 0.3595956566917491
261 971 -0.1734624626864245 -0.6999009108139718
261 981 -0.3130651914907041 0.14967430157983141
261 203 -0.3048171015033354 -0.24942689969922816
261 817 -0.15413492743468415 -0.2602873842439233
261 629 -0.32150282287912146 0.2374903935338546
261 38 -0.15567749698865807 0.533

264 518 -0.33614277304200346 0.2362403935338544
264 1133 -0.1876584399335588 -0.215842939799479
264 1101 -0.3474667914364349 -0.032509606466145424
264 473 -0.06708427649763848 -0.26432778828432735
264 1067 -0.19776869982119782 0.11008298612644696
264 819 -0.01903496789716641 0.21160804059267813
265 154 -0.43905286077358285 -0.5625096064661457
265 169 -0.495192407035578 -0.5871607692568435
265 499 -0.44686100540959917 0.06294493898839981
265 130 -0.2728839380169461 0.017490393533854398
265 146 -0.484366143777617 -0.7083160580790486
265 162 -0.46301574342303287 -1.7682238921804314
265 365 -0.3672024824283291 -0.7083160580790486
265 487 -0.5536910901606018 0.2280167093233283
265 401 -0.47480898501248436 -0.21057978190474191
265 192 -0.43905286077358285 -0.5734186973752364
265 244 -0.495192407035578 -1.204731828688368
265 112 -0.48815215125384226 -1.2602873842439233
265 718 -0.47768890697281075 -0.2825096064661454
265 676 -0.414471741113867 -0.1682238921804311
265 745 -0.5466463367365515 -

268 1011 -0.9105002559223903 -0.2773814013379403
268 290 -0.901517935258858 0.6147126157560767
268 966 -1.0564019644798264 0.1737403935338544
268 317 -1.086717501477081 0.5651094411529018
268 1129 -0.9364599450665967 0.18415706020052092
268 896 -0.9772447455989961 0.14635637291529768
268 1113 -0.9010544170674626 0.4237403935338544
268 1182 -0.8579752213908689 0.09643776195490705
268 803 -1.1457999185066607 -0.30603901823085167
268 1094 -0.9534815850720748 0.3263139229456189
268 1076 -0.9432761569888033 0.16966430657733245
268 891 -0.879082187111309 -0.08250960646614569
268 331 -1.1483737195532386 0.36285121827612254
268 341 -1.020844893687721 0.17658130262476357
268 343 -0.9364599450665967 0.345076600430406
268 1188 -0.9731073676271604 0.25662082831646327
268 1026 -0.962582296163101 0.47749039353385436
268 0 -0.9746950330239041 0.37360550144752347
268 174 -0.9117095348001132 -0.13922602437659348
268 703 -0.9141385805729976 0.08505796110142194
268 1123 -1.0564019644798264 0.105725687651

274 321 0.42530516167499366 0.5174903935338544
274 554 0.35934887605729227 -1.0088253959398297
274 766 0.2135612350418863 -0.0010281249846642027
274 665 0.595460953072307 0.2527845111809133
274 267 0.3803418465118673 0.594413470456931
274 1053 0.2514046826588321 0.2317761078195688
274 717 0.2514046826588321 0.3258237268671875
274 907 0.27535212411530674 0.24198018945222177
274 598 0.39400825383760285 -0.3942743123484984
274 575 0.11155763604048108 0.5280724041158646
274 386 0.6382786961403375 0.0799903935338544
274 203 0.07897674343459382 -0.24942689969922816
274 750 0.595460953072307 0.1246332506767116
274 1210 0.049990622408705665 0.6713365473800086
274 49 0.15733703857354442 0.13112675717021816
274 970 0.595460953072307 -0.10155722551376467
274 1178 0.14113299520723555 0.35749039353385426
274 787 0.4319678200145951 0.25662082831646327
274 1200 0.02537624582198772 0.39984333471032496
274 302 0.3013114737146805 0.4174903935338543
274 920 0.2015622498212788 0.4460618221052828
274 612 0

282 4 0.09821614458000782 0.3647126157560767
282 31 0.012643783340459347 -0.22609935005588921
282 816 0.1990483704406678 0.16454921706326608
282 725 0.01789849622733275 -0.8354507829367339
282 723 -0.542603871980678 0.24722012326358422
282 630 -0.008117784178182763 0.1674903935338543
282 502 -0.25187277341967373 -0.05393817789471722
282 359 0.37770544729729494 -0.600156665289675
282 468 -0.1684657131485574 -0.2902019141584531
282 846 -0.1394457342106561 -0.5972637048268012
282 218 0.15983965599051952 0.46193483797829904
282 429 -0.30663598272272896 0.4081153935338544
282 181 -0.08810312307189516 0.06294493898839981
282 111 0.05247019002738229 0.08891896496242602
282 619 0.03941282186437833 0.45951937904110096
282 6 -0.008117784178182763 -0.06761598944486913
282 320 0.0753154004342127 0.3674903935338545
282 777 0.04288285673454961 0.11749039353385449
282 298 0.01779248399988065 0.6407780647667312
282 300 0.28840199362586016 0.7397126157560767
283 101 -0.4970223915387084 -0.2463984953550

286 492 -0.1835061135546976 -0.09541283227259711
286 635 -0.028962740149451882 -0.31924430034369644
286 1135 -0.04080529488273177 0.3612403935338544
286 291 -0.0961798072331197 0.42658130262476357
286 412 0.021947121590077945 0.0336194257919189
287 1071 0.7113175540276813 0.24476312080658147
287 910 0.658135836185402 0.04380618300753847
287 1113 0.6709761006480126 0.4237403935338544
287 156 0.7992530181720015 0.5395492170632665
287 890 0.7807566332410834 0.2648588145864861
287 306 0.8333682988824962 0.6187562163186642
287 1205 0.804522684666606 0.13653801258147347
287 1064 0.6688750089184698 0.47749039353385436
287 217 0.8333682988824962 0.7026755787190395
287 1176 0.6745785161392175 0.2952681713116321
288 113 -0.04954103836370205 0.530823726867188
288 170 -0.007980164228189146 -0.16382828778482672
288 117 0.06452817925924492 -0.26552847439067406
288 137 0.0567020960937894 -0.062219751393681744
288 127 -0.08034874212040721 -0.45076357472011397
288 110 0.045889705065694564 -0.1336723971

295 569 0.0977946505756635 0.5636442396877004
295 393 -0.008461712793063736 -0.051954050910590244
295 850 -0.154631078727117 -0.042033415989954914
295 133 -0.02929675506052032 -1.0734186973752364
295 171 0.18152540742157397 0.2503148210147703
295 418 0.08203488878535214 0.4374903935338543
295 1090 -0.08257515605850685 0.42225229829575905
295 886 -0.1804875292514767 0.0545274305708916
295 117 0.0024828437235213663 -0.26552847439067406
295 302 -0.016465992758978754 0.4174903935338543
295 331 -0.0186403409967415 0.36285121827612254
295 472 -0.001741531100551974 -0.07267354089237532
295 781 -0.229445318965102 0.4103475363909972
295 600 -0.08775182645323287 0.005295271582634875
295 279 0.04007799030933379 0.605402481445942
295 198 0.04774836990611142 -1.1491762731328121
295 818 -0.12085888375124831 0.30127417731763817
295 402 -0.05205117600771114 -0.05393817789471722
295 45 0.01135749593037394 -0.6211234678522843
295 167 0.007524207379736136 0.0799903935338544
295 599 -0.02520856840385921 -

302 316 0.5354439048738446 0.6008237268671874
302 429 0.30460509532976854 0.4081153935338544
302 1208 0.31961414276807965 0.3508237268671879
302 1161 0.3323239589053246 -0.1075096064661456
302 1142 0.05115226460928701 0.2774903935338542
302 113 0.45461716277139397 0.530823726867188
302 1000 0.022719440521628886 -0.20045832441486366
302 816 0.5674893420312572 0.16454921706326608
302 984 0.3463810029797143 -0.3543044782610174
302 1096 0.4382175833668729 0.4374903935338543
302 335 0.14342185997310342 0.3699494099272971
302 1031 0.40113806345914416 0.09249039353385458
302 1147 0.5326970218050486 0.11008298612644696
302 878 0.39588187410287423 0.5174903935338544
302 993 0.560439884056477 0.2508237268671878
302 859 0.022719440521628886 0.4312834969821302
302 1128 0.17800001854776726 -0.12250960646614573
302 855 0.43357362004011835 -0.527964151920691
302 875 0.560439884056477 0.23543911148257246
302 896 0.006982207598956958 0.14635637291529768
302 790 0.18536837669190165 0.2243869452579923
30

317 195 -0.0947702033220974 0.017490393533854398
317 203 -0.07865502915558853 -0.24942689969922816
317 267 0.06986667966810245 0.594413470456931
317 957 0.03520970633989906 -0.3158429397994791
317 575 -0.04170476785305155 0.5280724041158646
317 496 -0.07438129752811529 -0.2950096064661456
317 422 -0.001375777469858619 0.1319482248591557
317 377 0.21856437921404437 -0.06432778828432761
317 946 0.0030939841677191104 -0.06028738424392355
317 518 0.112725938425561 0.2362403935338544
317 399 -0.04447462321344675 -0.2491762731328122
317 836 -0.16679411588419546 -0.019546643503182803
317 864 0.03520970633989906 -0.4825096064661456
317 174 0.0030939841677191104 -0.13922602437659348
317 761 0.09276064576927734 0.0603475363909971
317 394 -0.1156748123533613 0.08159295763641872
317 106 -0.2521710979398022 -0.5658429397994791
317 710 -0.04170476785305155 -0.004248736900928307
317 252 0.18566195109593275 0.22203584807930898
317 933 -0.08818051194306964 -0.16250960646614576
317 608 0.147803445508752

323 301 0.7345140688438874 0.7079665840100451
323 1142 0.627488623935553 0.2774903935338542
323 875 0.7345140688438874 0.23543911148257246
323 237 0.7446861135516882 0.21193483797829904
323 1083 0.7345140688438874 0.2821962758867955
323 322 0.5730069783619904 0.6603475363909976
323 38 0.702865284553909 0.5331153935338544
323 896 0.5950785619070031 0.14635637291529768
323 91 0.6639474922197643 0.4430223084274716
323 410 0.5730069783619904 0.42658130262476357
323 326 1.2174903935338546 0.5545274305708916
323 844 0.620525867018181 -0.15692821111730826
323 644 0.7112440939483036 0.10082372686718788
323 267 0.567440609265685 0.594413470456931
324 920 -0.07113236095716369 0.4460618221052828
324 32 -0.11702737296360732 -0.5175973257643913
324 99 0.4662083422518042 0.8616764400454824
324 447 -0.19098418273733175 -0.5658429397994791
324 396 -0.11702737296360732 -0.577747701704241
324 212 0.8508237268671883 -0.44679532075186
324 575 -0.2738765129409648 0.5280724041158646
324 187 -0.2825096064661

330 1004 -0.48375231748803715 0.09720053846139054
330 787 -0.21253760844275416 0.25662082831646327
330 78 0.020149968001939023 -0.14917627313281212
330 174 -0.3244176132804557 -0.13922602437659348
330 920 -0.2244808135536327 0.4460618221052828
331 738 0.9281813974451967 0.061934837978298685
331 925 1.0797083174539055 0.5394684155118763
331 371 0.6856063355628383 -0.5325096064661454
331 546 0.9539724782244074 -0.18521230916884823
331 642 0.8332554271999801 0.46343633947980045
331 197 1.03539672962201 0.3674903935338545
331 888 0.9969676055630754 -0.3731346064661456
331 502 0.872898782054826 -0.05393817789471722
331 827 0.872898782054826 0.4644600905035512
331 442 0.9640923352814288 0.3508237268671879
332 8 0.07795550981292498 0.19834145736364173
332 116 -0.014560055042402542 0.40824669605486275
332 7 -0.05268012233059327 0.25218427108487473
332 302 0.15607735005559187 0.4174903935338543
332 160 0.25239429929951074 0.46810767748447146
332 359 0.3728750089184705 -0.600156665289675
332 38 

340 998 0.4999317230152207 0.1374903935338545
340 1074 0.5185038900811221 0.37955935905109595
340 1207 0.4939313870736708 0.15749039353385452
340 306 0.5299918932339112 0.6187562163186642
340 1073 0.5101111054699401 -0.5325096064661454
340 355 0.45818147140904575 -0.13768202025924925
340 253 0.47679496155101475 0.34101980529856046
340 594 0.32468135167191825 -0.3753667493232884
340 47 0.5843617081224308 -0.776627253524969
340 983 0.6028580572226381 -0.8286634526199919
340 248 0.33804718806758505 0.38983081906576933
340 845 0.32101854110713424 0.032641908685369536
340 153 0.5205901560827164 -1.215842939799479
340 363 0.4336200225924589 -0.69679532075186
340 1105 0.47681172778658887 0.030310906354367173
340 668 0.5170484655024801 0.19140343701211515
340 875 0.5299918932339112 0.23543911148257246
340 394 0.5407804927026967 0.08159295763641872
340 125 0.43937885384165076 -0.14084293979947882
340 597 0.5085330859730801 -1.6115418645306618
340 722 0.5473624554201977 -0.5877727643608823
340 4

342 960 0.5250624502021508 -0.3039381778947172
342 1000 0.48630584617523187 -0.20045832441486366
342 1071 0.3335818960315806 0.24476312080658147
342 802 -0.15160814734162065 -0.6404043433082509
342 302 0.4422215763295534 0.4174903935338543
342 1007 0.3996635182966881 0.2902176662611273
342 827 0.5734301507184267 0.4644600905035512
342 805 0.40843113934499353 -0.7868574325531021
342 1198 0.4425827259938435 -0.18839195940732223
342 846 0.47151822988982284 -0.5972637048268012
342 821 0.48630584617523187 0.14086701691047798
342 480 0.6296883624844782 0.3905062665497274
342 935 0.5981122919864421 0.24883367711594406
342 592 0.5638524166039507 -0.3575096064661456
342 1173 0.5981122919864421 0.2317761078195688
342 866 0.5734301507184267 -0.1736860770543811
342 1014 0.17345872483714508 -0.2325096064661456
342 1195 0.40843113934499353 0.0487403935338544
342 426 0.4462508392163067 -0.28685743255310214
342 932 0.61405027387752 -0.11887324282978184
342 1094 0.5638524166039507 0.3263139229456189
34

346 507 -0.6138858931607678 0.11008298612644696
347 3 -0.1930359222556195 -0.42190354586008505
347 41 -0.1930359222556195 -0.039886655646473645
347 30 -0.1930359222556195 -1.0325096064661454
347 14 0.7674903935338544 0.17055161802365015
347 122 0.7602088401357965 -0.21327883723537644
347 195 0.6916055168802973 0.017490393533854398
347 207 0.8291787052221653 -0.3744014983580377
347 203 0.7317902942289884 -0.24942689969922816
347 29 0.9460618221052832 0.08611784451424676
347 360 0.7317902942289884 0.3143653935338544
348 72 -0.0295684299955572 -0.017393327396378222
348 523 0.05693235013047859 0.02692435579800545
348 285 0.0901399661834259 -0.05393817789471722
348 1070 0.5696230949556544 -0.7206048445613837
348 156 0.08921170500926445 0.5395492170632665
348 623 0.0901399661834259 -0.05393817789471722
348 991 0.15685873727237798 -0.19679532075186001
348 912 0.10098841724926899 -0.411081035037574
348 521 0.5021869634546983 0.42658130262476357
348 287 0.0901399661834259 0.617490393533854
348 

365 224 0.5551374523573833 0.49788255039659957
365 614 0.3974656461632633 0.08645591077523385
365 305 0.36495160650987746 0.19021766626112724
365 417 0.437011872598561 0.11749039353385449
365 200 0.437011872598561 0.6816694980114666
365 27 0.4135770483327552 0.5174903935338544
365 175 0.2886339084993197 0.6530836138728371
365 641 0.38879335116774794 0.3239420064370804
366 120 -0.1434615000178332 -0.2325096064661456
366 186 0.01982463112138877 -0.09312907549269411
366 100 -0.235388422039168 -0.8158429397994791
366 78 0.13978565510602703 -0.14917627313281212
366 99 0.05576894024665613 0.8616764400454824
366 25 0.03604445623378982 0.6548949736865262
366 123 -0.12133313587790973 -0.12286048365912805
366 114 -0.4593212006690446 0.017490393533854398
366 112 -0.300752217138875 -1.2602873842439233
366 170 -0.1403878528049476 -0.16382828778482672
366 29 0.12337274647503094 0.08611784451424676
366 151 -0.12363926787978174 -0.07074490058379279
366 79 -0.35750960646614605 -0.7047318286883679
366 6

369 655 0.36032241220038763 0.017490393533854398
369 149 0.2651392767533838 -0.1555865295430685
369 568 0.36032241220038763 0.08891896496242602
369 512 0.2467592536119496 0.4674903935338546
369 365 0.07420867422453448 -0.7083160580790486
369 643 0.06317531037302215 -0.7982990801503562
369 791 0.35546278948715493 -0.1311582551147943
369 992 0.18531421648893032 -0.2325096064661456
369 381 0.08356092149576488 -0.4200096064661456
369 850 0.08342605133508307 -0.042033415989954914
369 419 0.25451106883916186 0.2582311342745953
369 956 0.21908135083558378 0.345076600430406
369 237 0.21421309451956327 0.21193483797829904
369 700 0.30017256660683334 0.2049903935338544
369 278 0.02896307394878983 -0.11213923609577536
369 1099 0.06191612548039327 0.13287500891846982
369 516 0.27754009674121205 -0.5450096064661456
369 51 0.24533620126158473 -0.4480268478454561
369 831 0.30017256660683334 0.4237403935338544
369 645 0.09854614939149142 -0.03514118541351419
369 273 0.23924781416861718 0.1841570602005

374 38 0.2357873915934383 0.5331153935338544
374 217 0.3050502021462944 0.7026755787190395
374 253 0.12507416778429636 0.34101980529856046
374 934 0.0841279720291408 -0.5413331358779101
374 1055 0.2357873915934383 0.28672116276462356
374 1064 0.24921762914022683 0.47749039353385436
374 381 0.11390163048046364 -0.4200096064661456
374 744 0.11946999761303845 -0.2902019141584531
374 574 0.1909858100344537 0.38983081906576933
374 268 0.19396098176914967 0.42446713771990074
375 263 0.13774355809081573 0.3299903935338544
375 156 0.15302152906499078 0.5395492170632665
375 738 -0.023466699035833205 0.061934837978298685
375 925 0.25437124325517324 0.5394684155118763
375 975 0.07298104773946346 0.2206153935338544
375 702 0.2921473865659632 0.04380618300753847
375 868 0.2921473865659632 0.3612403935338544
375 335 0.15773704716598003 0.3699494099272971
375 145 -0.03343526703816391 0.282591064674794
375 575 -0.10384807129402507 0.5280724041158646
375 1048 0.16729933936437957 -0.0687165030178698
375

383 299 -1.11948504713286 0.6130786288279721
383 790 -1.1602566055427825 0.2243869452579923
383 832 -1.024452231516456 0.18415706020052092
383 896 -1.117192650993453 0.14635637291529768
383 652 -0.8339744599807926 -0.13367239716381984
383 201 -1.097226571621316 -1.2244450903371131
383 394 -1.151371708608651 0.08159295763641872
383 813 -0.7886859413847849 0.14606182210528296
383 353 -0.9676761662555395 -0.3534773484016296
383 508 -0.9677602192630328 -0.24441436837090746
383 622 -1.072154957440052 0.1737403935338544
383 717 -1.0552748627481434 0.3258237268671875
383 472 -1.1529687459405866 -0.07267354089237532
383 862 -3.4825096064661456 0.11208498812844914
383 210 -0.8665742715931666 -0.2756130547420077
383 939 -1.2007608209005665 -0.2425096064661454
383 307 -0.770747869969425 0.4329833512803334
383 455 -0.8471291506917145 0.33144388190594753
383 610 -1.2007608209005665 -0.4055865295430685
383 413 -1.0457287530241017 0.20041722280214724
383 192 -1.2007608209005665 -0.5734186973752364
38

391 186 -0.2315094582960473 -0.09312907549269411
391 298 0.5135794428720053 0.6407780647667312
391 327 0.46827372194059524 0.45191662304205105
391 19 0.4929324367558574 0.22203584807930898
391 284 0.5331784258063803 0.24825962430308524
391 377 0.4447899285230803 -0.06432778828432761
391 379 0.2957454668563262 -0.1075096064661456
391 8 0.4929324367558574 0.19834145736364173
391 275 0.42694823143065763 -0.2719832906766717
391 313 0.13956235682774398 0.3424903935338546
391 51 0.3342964468990508 -0.4480268478454561
391 80 -0.0783525625862369 0.6841570602005214
391 196 -0.14640042302871725 -0.17481729877383811
391 4 -0.13456732960884965 0.3647126157560767
391 318 0.577228527740075 0.3674903935338545
391 321 0.5891463208686916 0.5174903935338544
391 148 0.19662871552931938 0.2821962758867955
391 93 0.577228527740075 -0.13768202025924925
391 301 0.4929324367558574 0.7079665840100451
391 309 0.577228527740075 0.5683378511609729
391 122 -0.3147536151807433 -0.21327883723537644
391 203 -0.175884

399 16 0.2486731892327798 -0.5658429397994791
399 69 0.44154102644524684 -0.6700096064661456
399 78 0.30218858653654523 -0.14917627313281212
399 101 0.3448348282402507 -0.24639849535503444
399 68 0.3724944952565772 0.031003907047367996
399 121 -0.06082736347549256 -1.4825096064661456
399 47 0.40968411312695396 -0.776627253524969
399 92 0.3398944251396543 0.512705704538639
400 846 -0.41717613214750093 -0.5972637048268012
400 631 -0.39788891188615416 0.15385402989749064
400 1079 -0.09242227006801418 -0.1253667493232884
400 750 0.1026182136768834 0.1246332506767116
400 407 -0.4005135695149047 -0.3825096064661455
400 97 -0.4013790695625654 -0.09120525864005868
400 856 -0.3238958803354217 -0.6363557603122993
400 396 -0.34572733165491476 -0.577747701704241
400 745 -0.2408203381924463 -0.3825096064661455
400 748 0.1026182136768834 0.22119409723755812
400 440 -0.45503701078380576 -1.5325096064661456
400 379 -0.2096305570350636 -0.1075096064661456
400 175 -0.23841328452355715 0.6530836138728371

404 113 1.2118642244130373 0.530823726867188
404 21 1.05696407774438 -0.9673580913146305
404 408 1.074513202657506 0.3143653935338544
404 478 0.8872320909508287 -0.3325096064661457
404 172 1.0035347414947808 0.34755575301097874
404 370 1.0328835100479146 -0.2682238921804312
404 663 1.190801578472946 0.5293951554386163
404 817 1.0120254532278206 -0.2602873842439233
404 23 1.1407810860790666 0.4470678583225869
404 274 1.054092291410044 -0.41693583597434225
404 717 1.0559996247270207 0.3258237268671875
404 512 1.131327155912334 0.4674903935338546
404 787 1.131327155912334 0.25662082831646327
404 628 1.074513202657506 0.17463325067671143
404 761 1.0765348282869733 0.0603475363909971
404 765 1.0916525437384803 -0.8991762731328121
404 273 0.9771965033559753 0.18415706020052092
404 777 1.1947460740986071 0.11749039353385449
404 436 1.0559996247270207 -0.03806516202170096
404 52 1.0987214107036345 -0.3671249910815302
404 219 1.0304721379557615 -0.9825096064661456
404 296 1.0856967185903517 0.4

414 292 -0.9086380287303899 0.42225229829575905
414 191 -0.8806075961134305 -0.020971144927683927
414 340 -0.47066803847921346 0.6490693409022756
414 23 -0.6375920971526798 0.4470678583225869
414 304 -0.16136987771163458 0.4862403935338544
414 858 -0.7612306001298599 0.2049903935338544
414 312 -0.6043166253397882 0.5519731521545439
414 993 -0.35641761285372064 0.2508237268671878
414 1165 -0.9455412582960165 0.06749039353385422
414 1191 -0.5219642160339824 0.2032046792481399
414 14 -0.6369995693382111 0.17055161802365015
414 461 -0.6452601992663198 0.2187890948325557
414 591 -0.8433513559531289 -0.44250960646614557
414 156 -0.4592006845037844 0.5395492170632665
414 357 -0.7197707529629613 0.20170091984964378
414 71 -0.7769267569797647 -0.5866762731328121
414 393 -0.8433513559531289 -0.051954050910590244
414 89 -0.6792919332760845 -0.3902019141584532
414 710 -0.8605454559990919 -0.004248736900928307
414 80 -0.9552841307545785 0.6841570602005214
414 1139 -0.6910239790181034 -0.30603901823

420 648 0.2888162526613214 0.13653801258147347
420 303 -0.034898746737639375 0.365975242018703
420 484 0.22690826375148943 -0.1295684299955573
420 314 0.37364125864557574 0.6753851303759597
420 710 -0.24268101351078686 -0.004248736900928307
421 43 0.17980299558224244 0.2763556417608046
421 10 -0.028648966164826817 -0.8461459701025094
421 196 0.08361259145621647 -0.17481729877383811
421 68 -0.011336644239505311 0.031003907047367996
421 172 0.10741845108781067 0.34755575301097874
421 123 0.0508237268671885 -0.12286048365912805
421 174 -0.13421239284385544 -0.13922602437659348
421 78 0.1767356765527226 -0.14917627313281212
421 170 0.12273154212599291 -0.16382828778482672
421 25 -0.2439624860472982 0.6548949736865262
421 109 -0.037206396002412045 0.14825962430308515
421 187 0.1323237268671864 -0.3473744713310105
421 1 0.21907326131783877 -0.17871213811171538
421 5 0.15870014114047004 -0.3849486308563894
421 56 -0.13421239284385544 -0.5575096064661458
421 23 -0.22303438780725493 0.447067858

425 92 -0.09155413693504277 0.512705704538639
425 117 -0.267465823877834 -0.26552847439067406
425 913 -0.24491809208326298 -0.3916005155570548
425 525 -0.34692520439543895 0.3174903935338542
425 125 -0.1646154526706578 -0.14084293979947882
425 291 -0.26586361571391537 0.42658130262476357
425 805 -0.14005619063313413 -0.7868574325531021
425 1007 -0.012533600067856643 0.2902176662611273
425 1018 0.010859581517953742 -0.3916005155570548
425 262 -0.24491809208326298 0.3950414139420175
425 462 0.05320467924814043 0.34101980529856046
425 418 0.0033121074614816948 0.4374903935338543
425 395 -0.14005619063313413 -0.9696890936456328
425 101 -0.08579060840653474 -0.24639849535503444
425 612 -0.34692520439543895 0.2774903935338542
425 268 -0.02229698226637744 0.42446713771990074
426 394 0.6877073458457699 0.08159295763641872
426 1104 0.6646236187974055 -0.06359068754722674
426 1144 0.6676193627998246 0.3323052083486693
426 499 0.7382426433607914 0.06294493898839981
426 886 0.670845121946321 0.054

429 820 0.4549903935338544 -0.7741762731328121
429 590 0.5362208826077377 -0.22609935005588921
429 808 0.532178276030796 -0.4825096064661456
429 22 0.43969276481856934 -0.16432778828432726
429 40 0.5643965811586047 -0.4825096064661456
429 828 0.5102440167222602 0.40121132376641233
429 688 0.5102440167222602 -0.14917627313281212
429 473 0.10263325067671047 -0.26432778828432735
429 137 0.22533353078875473 -0.062219751393681744
429 175 -0.7007837181412726 0.6530836138728371
429 810 0.5643965811586047 -0.8461459701025094
429 331 0.4549903935338544 0.36285121827612254
429 170 0.39836733110343703 -0.16382828778482672
429 395 -0.6106765215183065 -0.9696890936456328
429 224 -0.9547318286883675 0.49788255039659957
429 571 -0.2220305645499785 0.41404211767178545
429 689 0.5670673369500343 0.24476312080658147
429 366 -0.8342537925126581 -0.215842939799479
429 655 0.4862403935338544 0.017490393533854398
429 305 0.3049260269015166 0.19021766626112724
429 691 0.5769597439455674 0.1508237268671877
42

430 290 -0.501149762688406 0.6147126157560767
430 940 -0.25071625936110964 -0.1253667493232884
430 802 -0.49460704721874515 -0.6404043433082509
430 961 -0.17792429246219044 -0.14917627313281212
430 705 -0.25071625936110964 0.0545274305708916
430 327 -0.07219428014007301 0.45191662304205105
430 892 -0.62075668153008 0.261080137123598
430 325 -0.49460704721874515 0.04129991734337812
430 641 -0.2793523196792722 0.3239420064370804
430 350 -0.08540840838135688 -0.30603901823085167
430 1018 -0.11532165563229446 -0.3916005155570548
430 734 -0.5162259657041264 -0.4480268478454561
430 343 -0.5472318694003384 0.345076600430406
430 146 -0.5299081095280687 -0.7083160580790486
430 819 -0.4910116256957173 0.21160804059267813
430 89 -0.3813240416943615 -0.3902019141584532
430 297 0.13717719442870324 0.43177610781956854
430 109 -0.21453083040156828 0.14825962430308515
430 1001 -0.31112459465596665 -0.3212192838855006
430 992 -0.21453083040156828 -0.2325096064661456
430 591 -0.5437502690189282 -0.44250

433 810 0.30915706020052314 -0.8461459701025094
433 1033 0.17006225435080324 0.3081880679524591
433 392 0.30915706020052314 -0.8575096064661456
433 633 -0.12003927599325825 0.6532046792481405
433 145 -0.34715634745513446 0.282591064674794
433 740 0.30915706020052314 -0.3625096064661455
433 751 0.30915706020052314 -0.21385289004823527
433 23 -0.4817784787015684 0.4470678583225869
433 491 0.589960809092061 -0.22325034720688652
433 814 0.30915706020052314 -0.08857021252675157
433 977 0.15056716080142563 0.10572568765150159
433 739 -0.2280055988185241 -0.4200096064661456
433 616 0.2338079369758912 -0.5539381778947172
433 1129 0.30915706020052314 0.18415706020052092
433 447 -0.24562106411552165 -0.5658429397994791
433 375 -0.5715313114048071 0.3214119621613052
433 771 0.30044800643237357 -0.7651183021183194
433 163 0.8538202137446427 -0.08857021252675157
433 1019 0.19929401358831766 -0.6930359222556195
433 1109 0.19929401358831766 -0.0342337443971803
433 674 0.30915706020052314 -0.220604844

439 705 -0.5514849404637459 0.0545274305708916
439 1044 -0.4430099152987581 -0.19679532075186001
439 426 -0.3287204186112942 -0.28685743255310214
439 1132 -0.5574930902802837 0.053204679248139986
440 790 0.6508579895765112 0.2243869452579923
440 1032 0.7216141048740607 0.19834145736364173
440 923 0.39805245442378556 -0.3100958133626972
440 971 0.3774903935338534 -0.6999009108139718
440 389 0.3774903935338534 -0.5641422595273702
440 984 0.39805245442378556 -0.3543044782610174
440 973 0.39805245442378556 -0.6894061581902835
440 1048 0.39615966946340286 -0.0687165030178698
440 751 0.3774903935338534 -0.21385289004823527
440 1098 0.9446412798287089 0.2424903935338545
440 1086 0.3774903935338534 -0.16250960646614576
440 34 0.3774903935338534 -0.1163124233675541
440 1154 0.5168396343147674 -0.6491762731328121
440 910 0.4272771622599354 0.04380618300753847
440 117 0.5387437995284032 -0.26552847439067406
440 1014 0.4066102826537441 -0.2325096064661456
441 5 0.48180857535203625 -0.3849486308563

458 47 0.14288924484208643 -0.776627253524969
458 110 0.6218453647746114 -0.13367239716381984
458 1 -0.21565567388187645 -0.17871213811171538
458 179 -0.024269071944988063 -0.05542627313281212
458 43 0.19352713163571833 0.2763556417608046
458 132 0.3634574828243875 -0.4825096064661456
459 200 0.041906529414959426 0.6816694980114666
459 263 0.017490393533855286 0.3299903935338544
459 844 -0.06992219387873222 -0.15692821111730826
459 351 -0.2825096064661463 0.48891896496242593
459 80 -0.1082079304884922 0.6841570602005214
459 234 0.25696935144968647 0.1303936193403059
459 175 0.4703205822131 0.6530836138728371
459 236 0.6536160479841153 0.6127284887719493
459 871 -0.11486254764261616 0.3174903935338542
459 326 -3.4825096064661456 0.5545274305708916
460 745 0.07385460314749492 -0.3825096064661455
460 12 0.16759285255024992 -0.38727151122805026
460 614 0.14524131351365144 0.08645591077523385
460 697 0.08685523611845314 -0.06946612820527598
460 903 0.23705496699357154 0.3508237268671879
460

466 569 0.040811289056242206 0.5636442396877004
466 542 -0.05490937773496496 -0.6729857969423363
466 1099 0.18812706461163797 0.13287500891846982
466 368 0.10015827270064026 0.0799903935338544
466 1090 0.15491630377940835 0.42225229829575905
466 1096 0.172352498061616 0.4374903935338543
466 155 0.16516655545289494 -0.047726997770493274
466 1095 0.040811289056242206 0.07681242743215932
466 760 0.15818116525538883 -0.33357343625337954
466 574 0.11754559640440387 0.38983081906576933
466 649 -0.1771876660772569 -0.059432683389222696
466 367 0.14124454635777584 -0.8509306590977244
466 1144 0.09494567534283149 0.3323052083486693
466 1135 0.04782544775927988 0.3612403935338544
466 1113 0.19697064932963482 0.4237403935338544
466 99 0.20068301085174278 0.8616764400454824
466 1006 0.20952305740496513 0.0658774903080479
466 1187 0.05899436707421124 0.3174903935338542
466 424 0.05065396816705636 0.4374903935338543
466 299 0.05899436707421124 0.6130786288279721
466 853 0.20068301085174278 -0.111081

472 175 0.2334100225910407 0.6530836138728371
472 1135 -0.08001710704311327 0.3612403935338544
472 848 0.12393343571253101 0.12999039353385422
472 113 -0.007338373589433456 0.530823726867188
472 575 0.0959812868556229 0.5280724041158646
472 1056 0.05307543575339668 0.36364423968770065
472 301 0.05307543575339668 0.7079665840100451
472 1175 0.09762942034616717 0.15851603455949537
472 1142 -0.025130247536943262 0.2774903935338542
472 1160 -0.02978093621249478 -0.03806516202170096
472 346 0.009758434770968716 0.365975242018703
472 890 -0.4113941656478306 0.2648588145864861
472 307 -0.024678281164940863 0.4329833512803334
473 80 0.5841015509027128 0.6841570602005214
473 187 0.37999039353385466 -0.3473744713310105
473 40 -1.300691424647964 -0.4825096064661456
473 198 -1.300691424647964 -1.1491762731328121
473 184 -1.300691424647964 -0.5625096064661457
473 196 0.3773996219756097 -0.17481729877383811
473 0 -0.6575668147041314 0.37360550144752347
474 139 0.5031019043252218 0.3896708446616741
4

490 872 1.038766989278535 0.11749039353385449
490 977 1.038766989278535 0.10572568765150159
491 245 0.5383804266342609 -0.17216477887993875
491 288 0.6906388548121778 0.6769106833889267
491 177 0.5490478744657152 0.14670387667992202
491 319 0.9931586594694282 0.48112675717021824
491 240 0.8415514870186915 0.4424903935338542
491 156 0.9698787794220833 0.5395492170632665
491 599 0.9177021457837191 -1.025987867335711
491 575 0.9615638492934546 0.5280724041158646
491 251 0.35394859579047155 -0.2902019141584531
491 393 0.5319966237737641 -0.051954050910590244
491 180 0.944170544310722 0.5748674427141824
491 301 0.9500364951191007 0.7079665840100451
491 298 0.9414408408841628 0.6407780647667312
491 305 0.6261891713915122 0.19021766626112724
491 271 0.9163426173502116 -0.042509606466145655
491 266 0.9753671673797348 -0.4454725694291084
491 297 0.9163426173502116 0.43177610781956854
491 217 0.9500364951191007 0.7026755787190395
491 175 1.0394303473444793 0.6530836138728371
491 666 0.8674903935

497 850 0.049408962767285125 -0.042033415989954914
497 1051 -0.012581452996031306 0.0799903935338544
497 470 -0.23879408464939678 -0.645300304140564
497 1062 0.01528260132606274 -0.25523687919341853
497 1054 -0.15462290387887778 -0.19679532075186001
497 450 -0.36058265727392014 0.07154444758790834
497 25 0.1338309540838285 0.6548949736865262
497 270 0.1338309540838285 -0.4522065761631153
497 66 -0.26428530740072187 0.2187890948325557
497 202 -0.09778605118836259 0.3587602348036958
497 413 -0.17434316462740185 0.20041722280214724
497 864 0.08857095730625675 -0.4825096064661456
497 1001 -0.03398567945215225 -0.3212192838855006
497 1058 -0.2658525128417941 0.029118300510598605
497 883 -0.03327821488038518 -0.6999009108139718
497 279 -0.1809327545272139 0.605402481445942
497 443 -0.23318901439203055 -0.1031992616385593
497 821 -0.13711158921970945 0.14086701691047798
497 1089 0.0327301101575026 0.3674903935338545
497 1020 -0.13557805568627757 -0.4825096064661456
497 1090 -0.049774735277562

506 356 0.24793964478177166 -0.10155722551376467
506 414 0.18650172608563276 -1.5241762731328123
506 536 0.03805032354260218 0.18415706020052092
506 651 -0.5498172987738381 -0.14917627313281212
507 532 0.7236337382437532 0.28672116276462356
507 590 0.4699455766983127 -0.22609935005588921
507 637 0.5541556869685604 -0.026627253524968975
507 333 0.8082243124910646 0.7079665840100451
507 349 0.9846794238962429 0.3924903935338544
507 340 0.9627830143989953 0.6490693409022756
507 290 0.5334903935338526 0.6147126157560767
507 214 0.6720039927329204 0.3174903935338542
507 346 0.5334903935338526 0.365975242018703
507 224 0.5401441152167012 0.49788255039659957
507 504 0.6720039927329204 -0.2920334159899549
507 527 0.6737172739284296 -0.21421692353931654
507 11 0.5971485131919754 -0.1432238921804312
507 156 0.9846794238962429 0.5395492170632665
507 139 0.5410594171028777 0.3896708446616741
507 113 0.7152952248222006 0.530823726867188
507 93 0.9627830143989953 -0.13768202025924925
507 477 0.79195

512 320 0.1652094160574964 0.3674903935338545
512 911 -0.12922762159804613 -0.09120525864005868
512 351 0.16978924410856555 0.48891896496242593
512 1009 0.1391898548947439 -0.01192137117202785
512 29 -0.1742110436474178 0.08611784451424676
512 966 -0.1742110436474178 0.1737403935338544
512 792 0.01687613112863451 -0.9491762731328124
512 426 -0.10285315648298798 -0.28685743255310214
512 442 -0.20995219878556215 0.3508237268671879
512 136 0.10222234553283727 -0.597894221850761
512 6 0.1391898548947439 -0.06761598944486913
512 824 -0.3692483657148635 0.6802810912082733
512 410 0.24088055341372527 0.42658130262476357
512 277 0.14700931797140848 -0.2325096064661456
512 846 0.10172518018755561 -0.5972637048268012
512 388 0.15421035685234896 -0.4055865295430685
512 313 -0.1744006851482598 0.3424903935338546
512 424 -0.10285315648298798 0.4374903935338543
512 993 0.025026554953572866 0.2508237268671878
512 890 -0.2707881613772489 0.2648588145864861
512 813 0.11484317453448956 0.146061822105282

522 727 -0.413368698472909 -0.042509606466145655
522 429 -0.4771663927375962 0.4081153935338544
522 760 -0.12283786820405052 -0.33357343625337954
522 59 -0.4041813738724205 -0.8528799768365158
522 367 -0.42644357946988665 -0.8509306590977244
522 999 -0.09321963502922248 0.18415706020052092
522 633 -0.24287741459208778 0.6532046792481405
522 326 -0.402964151920691 0.5545274305708916
522 279 -0.49261593153781513 0.605402481445942
522 426 -0.10771212671817487 -0.28685743255310214
522 814 -0.6910757144735973 -0.08857021252675157
522 1040 -0.4788426615619934 -0.21421692353931654
522 290 -0.4192713664993022 0.6147126157560767
522 591 -0.502308933985609 -0.44250960646614557
522 721 -0.4872810685784703 -0.4454725694291084
522 116 -0.12397698068957919 0.40824669605486275
522 653 -0.43402781268064405 -0.17481729877383811
522 94 -0.4363861242504994 -0.6075096064661456
522 1004 -0.4265791244435855 0.09720053846139054
522 668 -0.5028769283731211 0.19140343701211515
522 804 -0.5028769283731211 0.117

528 409 0.7206433674745729 -0.7456675012029876
528 489 0.6863242520228874 -0.017393327396378222
528 299 0.2748716783045526 0.6130786288279721
528 139 0.3317027223009794 0.3896708446616741
528 66 0.7206433674745729 0.2187890948325557
528 501 0.5919493300395202 -0.3991762731328121
528 132 0.2885571984714641 -0.4825096064661456
528 389 0.7206433674745729 -0.5641422595273702
528 610 0.7206433674745729 -0.4055865295430685
528 598 0.6512483553172941 -0.3942743123484984
528 295 0.7860306391000234 0.7527845111809133
528 202 0.759845892375957 0.3587602348036958
528 166 0.4800503935338565 0.0799903935338544
528 75 1.5174903935338544 0.8508237268671874
528 318 0.9522497518226296 0.3674903935338545
528 221 0.6518028159076326 0.4852323290177254
528 432 0.6512483553172941 -0.39679532075185975
528 451 0.8397334844917324 0.6227535514285916
528 265 0.5390650643099817 0.5174903935338544
528 424 0.8024177846204856 0.4374903935338543
528 417 0.7516183734585717 0.11749039353385449
528 480 0.600315041737924

529 506 -0.7150725214311251 0.5651094411529018
529 418 -0.6183417384539736 0.4374903935338543
529 636 -0.6857640741863191 -1.3713984953550344
529 837 -0.738957887622373 -0.6325096064661455
529 445 -1.0232049330199793 -1.0088253959398297
529 517 -0.8254997971276126 0.5174903935338544
529 682 -0.6541929803897664 0.4174903935338543
529 901 -0.7749710304193984 -0.16672013278193498
529 335 -0.8449725499930087 0.3699494099272971
529 944 -1.0232049330199793 0.07304594908940976
529 634 -0.8525210473437275 0.338918964962426
529 121 -0.9932392202000528 -1.4825096064661456
529 367 -0.7023562700693131 -0.8509306590977244
529 943 -0.7054131033784881 0.026262323358415607
529 613 -1.0232049330199793 -0.008825395939829672
529 853 -0.5528133380253024 -0.11108103503757416
529 42 -0.7449884249857615 -0.2325096064661456
529 857 -0.6318831518231445 -0.11408855383456684
529 159 -0.8254997971276126 -0.0645991587049517
529 47 -0.7004488287730841 -0.776627253524969
529 241 -0.7647955102604138 0.149069340902275

532 156 -0.15203660521039097 0.5395492170632665
532 137 -0.063668840497761 -0.062219751393681744
532 90 0.06260116668523796 -0.12044064094890405
532 37 -0.018538143921554173 0.4356190485046145
532 139 0.07687736671393175 0.3896708446616741
532 48 -0.059805670356619434 0.14415706020052088
532 60 0.04383881914448118 -0.38727151122805026
532 53 0.19719936685640915 -0.7825096064661454
532 92 -0.019707251278366922 0.512705704538639
532 97 0.1920277260163057 -0.09120525864005868
533 570 -0.5097025551594241 -0.14917627313281212
533 286 -0.3043548241728189 0.1424903935338544
533 64 -0.08977311997966231 0.2508237268671878
533 38 0.1192660008300317 0.5331153935338544
533 681 -0.03379445062052078 -0.9325096064661458
533 8 0.0564801692091117 0.19834145736364173
533 312 -0.06697548220301552 0.5519731521545439
533 614 -0.18894316632223163 0.08645591077523385
533 233 -0.5235447564177726 0.3239420064370804
533 31 -0.14050455747380441 -0.22609935005588921
533 224 0.21462669415057967 0.49788255039659957

536 890 -0.2660386262218557 0.2648588145864861
536 417 -0.20722654855382583 0.11749039353385449
536 824 -0.15092232421495932 0.6802810912082733
536 288 -0.39651612112412593 0.6769106833889267
536 1028 0.059330467880283067 0.4312834969821302
536 478 -0.30913305915855815 -0.3325096064661457
536 170 0.0028419327545328166 -0.16382828778482672
536 850 -0.15485534624974395 -0.042033415989954914
536 745 -0.21780209295563102 -0.3825096064661455
536 99 0.1483076112506354 0.8616764400454824
536 509 -0.2206121705687094 -0.2825096064661454
536 611 -0.29612372347930327 0.4122272356391177
536 700 0.006331576542807582 0.2049903935338544
536 237 -0.025518437953204298 0.21193483797829904
536 871 -0.19939583401105532 0.3174903935338542
536 554 -0.04192280940014337 -1.0088253959398297
536 1017 -0.07692700706830546 0.02860150464496547
536 455 0.16905040264339366 0.33144388190594753
536 694 0.16905040264339366 -0.16108103503757398
536 698 0.16905040264339366 0.3595956566917491
536 336 0.0012257841723477547

538 859 -0.47410676053876344 0.4312834969821302
538 205 -0.5043971811521679 -0.9682238921804314
538 892 -0.21025221637380254 0.261080137123598
538 59 -0.18422178985371973 -0.8528799768365158
538 738 -0.3012701850166901 0.061934837978298685
538 440 -0.1555088037510024 -1.5325096064661456
538 718 -0.6714012366394666 -0.2825096064661454
538 598 -0.8165459378795292 -0.3942743123484984
538 98 -0.5639248298025246 -0.3552368791934182
538 447 -0.24879970994107126 -0.5658429397994791
538 66 -0.8414151172703956 0.2187890948325557
538 150 -0.22004473292138815 -0.07074490058379279
538 776 -0.1920247106513071 0.2097980858415469
538 793 -0.8019093921038762 -0.07341869737523643
538 507 -0.6382569794020236 0.11008298612644696
538 376 -0.46726999571886285 -0.5182238921804312
538 402 -0.2535489142080225 -0.05393817789471722
538 767 -0.8165459378795292 -0.7444143683709075
538 142 -0.8414151172703956 0.017490393533854398
538 833 -0.6197364709373123 -0.21935171172930357
538 387 -0.5477835904874855 0.282196

In [176]:
better_MAE, better_RMSE = evaluation(better_result, test_user_dict, test_counter)    
print(better_MAE, better_RMSE)  # 0.6917506759527966 0.8983386235087089
# 0.7476839548925464 0.9730766472552652

0.6834455169311903 0.8918294879508567


Please explain what you do to improve the recommendation in 1-2 paragraphs.

The recommendation is improved mainly because I introduced extra information from movie info. When calculating the user bias. I came up with the idea that if current movie is similar with a movie that this user has rated before, that movie should have more weight when calculating the user bias. Since movie genres are specific terms, Jaccard Similarity is a good choice in this case. So I used Jaccard to measure the similarities between two movies. And if a movie has higher similarity, I will give it a higher weight in calculating the user bias. This method has a better MAE and is way more efficient than matrix factoriztion. 

## Collaboration declarations

*If you collaborated with anyone (see Collaboration policy at the top of this homework), you can put your collaboration declarations here.*

I found these websites useful when doing Part 2:  
https://datasciencemadesimpler.wordpress.com/tag/alternating-least-squares/  
http://www.quuxlabs.com/blog/2010/09/matrix-factorization-a-simple-tutorial-and-implementation-in-python/  
Collaborating with three classmates:  
Peixin Liu  
Yiran Huang  
Yicheng Li